[RFC 0049] Flakes (#49)
NixOS/rfcs: rfcs/0049-flakes.md

  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. This RFC,
however, only describes the format and semantics of flakes; it doesn't
describe changes to the `nix` command to support flakes.

# Detailed design
[design]: #detailed-design

## Flakes

A flake is a directory that contains a file named `flake.nix` in the
root directory. `flake.nix` specifies some metadata about the flake
such as dependencies (called *inputs*), as well as its *outputs* (the
Nix values such as packages or NixOS modules provided by the flake).

As an example, below is the `flake.nix` of
[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
for automatically fetching DWARF debug symbols by ELF build ID). It
depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  edition = 201911;

>   description = "A filesystem that fetches DWARF debug info from the Internet on demand";

Reviewer: Should we have an explicit `name` attribute like with derivations?

Author: There was a `name` attribute, but I got rid of it because it wasn't used anymore. (Originally it was used to determine the names of the arguments to the `outputs` function, but now those are determined by the `inputs` attrset.)

Reviewer: Isn't having a name for talking of a flake in general (and not in terms of a specific URL) useful on its own? For example, when people inevitably end up fetching the flake sources into the store…

Author: Maybe, but we can add such an attribute when there is a use case. Without that we can only speculate about what the semantics should be.

Reviewer: OK, I guess we could continue once the «download-all-dependencies» use case is addressed fully.

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

    nixosModules.dwarffs = ...;

    defaultPackage = packages.dwarffs;

    checks.build = packages.dwarffs;
  };
}
```

A flake has the following attributes:

* `edition`: A number that specifies the version of the flake
  syntax/semantics to be used. This allows the interpretation of
  flakes to change in the future. It also enables some evolution of
  the Nix language; for example, the Nix files in the flake could be
  parsed using a syntax determined by the edition. The only currently
  allowed value is `201911`. Nix rejects flakes with an unsupported
  edition.

* `description`: A short description of the flake.

* `inputs`: An attrset specifying the dependencies of the flake

>   (described below).

Reviewer: Is it permissible to have some of the input attributes be computed based on the other inputs?

* `outputs`: A function that, given an attribute set containing the
  outputs of each of the input flakes keyed by their identifier,
  yields the Nix values provided by this flake. Thus, in the example
  above, `inputs.nixpkgs` contains the result of the call to the
  `outputs` function of the `nixpkgs` flake, and
  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
  attribute of `nixpkgs`.

>   In addition to the outputs of each input, each input in `inputs`

Reviewer 2: Some of this metatdata seems to be quite input-type dependent (e.g. `rev`), whereas others seem like they'll be valid for any input type (e.g. `outPath`). Could we only include the generic ones here, and put the type-specific metadata along with the descriptions of the input types?

  also contains some metadata about the inputs. These are:

  * `outPath`: The path in the Nix store of the flake's source
    tree. This means that you could import Nixpkgs in a more
    legacy-ish way by writing

>         with import inputs.nixpkgs { system = "x86_64-linux"; };

Reviewer 2: Later, you say that access to files outside the current top-level flake are prohibited: isn't this a case of importing a file that's from a dependency (and hence outside the current top-level flake)?

Reviewer 3: @edolstra should this be `with import inputs.nixpkgs.outPath` ?

Author: No, you can import an attrset that contains `outPath` directly.

    since `nixpkgs` still contains a `/default.nix`. In this case we
    bypass its outputs entirely and only use the flake mechanism to
    get its source tree.

  * `rev`: The commit hash of the flake's Git repository.

  * `revCount`: The number of ancestors of the revision `rev`. This is
    not available for `github` repositories (see below), since they're
    fetched as tarballs rather than as Git repositories.

  * `lastModified`: The commit time of the revision `rev`, in the
    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
    this is available for both Git and GitHub repositories, so it's
    useful for generating (hopefully) monotonically increasing version
    strings.

  The value returned by the `outputs` function must be an attribute
  set. The attributes can have arbitrary values; however, some tools
  may require specific attributes to have a specific value (e.g. the
  `nix` command may expect the value of `packages` to be an attribute
  set of derivations).

## Flake inputs

The attribute `inputs` specifies the dependencies of a flake. These

> specify the location of the dependency, or a symbolic flake identifier

Reviewer: Will it be possible to force-override an input location (with the user being responsible for validity of such subsitution)?

Author: Yes.

Reviewer: Maybe mention it somewhere?

Author: Well, I removed all user interface aspects from the RFC :-) But it's something like `--override-flake OLD NEW`.

Reviewer: Ah, I underestimated the completeness of removal of UI mentions. I thought it is about «how something is done», but some of the «what should be possible to do» is still specified.

    inputs.import-cargo = {
      type = "github";
      owner = "edolstra";
      repo = "import-cargo";
    };

    # An indirection through the flake registry.
    inputs.nixpkgs.id = "nixpkgs";

Each input is fetched, evaluated and passed to the `outputs` function
as a set of attributes with the same name as the corresponding
input. The special input named `self` refers to the outputs and source
tree of *this* flake. Thus, a typical `outputs` function looks like
this:

    outputs = { self, nixpkgs, import-cargo }: {
      ... outputs ...
    };

It is also possible to omit inputs entirely and *only* list them as
expected function arguments in `outputs`. Thus,

    outputs = { self, nixpkgs }: ...;

without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in
the flake registry.

Repositories that don't contain a `flake.nix` can also be used as
inputs, by setting the input's `flake` attribute to `false`:

    inputs.grcov = {
      type = "github";
      owner = "mozilla";
      repo = "grcov";
      flake = false;
    };

    outputs = { self, nixpkgs, grcov }: {
      packages.grcov = stdenv.mkDerivation {
        src = grcov;
        ...
      };
    };

> The following input types are specified at present:

Reviewer 4: Specified where? This doesn’t go into any more detail, and they’re never mentioned again.

Reviewer 2: It would be really nice to be able to define new input/fetcher types. This would be somewhat tricky, since most such definitions would rely on a nixpkgs, so you need a kind of overlay-style process where the fetchers and inputs are mutually defined (see https://github.com/nmattia/niv/pull/137 for a POC implementation of such a scheme in `niv`).

However, otherwise I would guess that every fetcher needs to be implemented in the flake tool (`nix`, presumably), which makes it *much* harder to add, tweak, or improve them.

* `git`: A Git repository or dirty local working tree.

> * `github`: A more efficient scheme to fetch repositories from GitHub

Reviewer 4: Now that GitHub’s ETags have changed, why is this still separate?

Author: It's separate because the github scheme is more efficient in most cases but also provides less metadata (in particular `revCount`).

Reviewer 4: How is it more efficient than `tarball`?

Author: It isn't - it's more efficient than `git`. 

Note that `tarball` wouldn't allow you to specify `https://github.com/owner/repo/archive/master.tar.gz` and have it be locked properly. That is, the lockfile would record the content hash of `master.tar.gz`, but that will get invalidated.

  as tarballs. These have slightly different semantics from `git`
  (in particular, the `revCount` attribute is not available).

> * `tarball`: A `.tar.{gz,xz,bz2}` file.

Reviewer: Is this local, URL, or URL that can also be file:// URL?

* `path`: A directory in the file system. This generally should be

>   avoided in favor of `git` inputs, since `path` inputs have no

Reviewer 4: ```suggestion
  avoided in favor of `git` or `hg` inputs, since `path` inputs have no
```

  concept of revisions (only a content hash) or tracked files
  (anything in the source directory is copied).

* `hg`: A Mercurial repository.

## Lock files

Inputs specified in `flake.nix` are typically "unlocked" in that they
don't specify an exact revision. To ensure reproducibility, Nix will

> automatically generate and use a *lock file* called `flake.lock` in

Reviewer: If a flake contains a locked (already in `flake.nix`) input, should there be an option to use that revision for all unlocked instances of the same input?

              "rev": "c33e13881386931038d46a7aca4c9561144d582e"
            }
        },
        "nixpkgs": {
            "inputs": {},
            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
            "original": {
              "type": "github",
              "owner": "NixOS",
              "repo": nixpkgs"
            },
            "resolved": {
              "type": "github",
              "owner": "NixOS",
              "repo": "nixpkgs",
              "rev": "4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
            }
        }
    }
}
```

Thus, when we build this flake, the input `nixpkgs` is mapped to
revision `c33e13881386931038d46a7aca4c9561144d582e` of the
`edolstra/import-cargo` repository on GitHub. Nix will also check that
the content hash of the input is equal to the one recorded in the lock
file. This check is superfluous for Git repositories (since the commit
hash serves a similar purpose), but for GitHub archives, we cannot
directly check that the contents match the commit hash.

Note that lock files are only used at top-level: the `flake.lock`
files in dependencies (if they exist) are ignored. The lock file

> transitively locks direct as well as indirect dependencies.

Reviewer 4: It would be useful to have an option when fetching a flake as input to use the built version of that flake according to its own lockfile. But that can be a future enhancement.

## Reproducible evaluation

Lock files are not sufficient by themselves to ensure reproducible
evaluation. We also need to disallow certain impurities that the Nix
language previously allowed. In particular, the following are

> disallowed in a flake:

Reviewer 5: Flakes for the most part feels like an attempt to solve the centralisation problem: it's too difficult / awkward to manage things outside of the nixpkgs monorepo. I can see how improving on reproducability is important, but why must it be inextricably linked with flakes?

These impurities have allowed existing tools (like my own [nix-wrangle](https://github.com/timbertson/nix-wrangle/)) to serve as testing grounds for similar approaches to flakes, which lets us try out new ideas before baking them into nix.

It seems reasonable for flakes to be able to lock down these features - i.e. to have an attribute which opts in or out of this strict reproducible mode, and have that be enforced by the evaluator. A strict flake couldn't depend on a non-strict flake, and there could potentially be a toplevel command line argument to `nix` to treat the toplevel environment as strict/lax, a bit like `build-use-chroot` and similar settings.

Reviewer: I am not sure flakes are going to solve the problems with packages maintained outside Nixpkgs at scale, so maybe reproducible evaluation is the only thing that has a chance of working with flakes.

> * Access to files outside of the top-level flake or its inputs, as

Reviewer 2: Reading this, it's not clear to me whether I'd be able to do things like 
```
packages.foo = inputs.nixpkgs.callPackage ./foo.nix {};
```
Is this "inside" the current flake?

Author: Yes, `./foo.nix` is inside the current flake.

  well as paths fetched using `fetchTarball`, `fetchGit` and so on
  without a commit hash or content hash. In particular this means that
  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.

* Access to the environment. This means that `builtins.getEnv "<var>"`
  always returns an empty string.

* Access to the system type (`builtins.currentSystem`).

* Access to the current time (`builtins.currentTime`).

* Use of the Nix search path (`<...>`); composition must be done
  through flake inputs or `fetchX` builtins.

# Drawbacks
[drawbacks]: #drawbacks

Pure evaluation breaks certain workflows. In particular, it breaks the
use of the Nixpkgs configuration file. Similarly, there are people who
rely on `$NIX_PATH` to pass configuration data to NixOps
configurations.

# Alternatives
[alternatives]: #alternatives

For composition of multi-repository projects, the main alternative is
to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
in other repositories. However, since there is no explicit listing of

> dependencies, this does not provide automatic updating.

Reviewer 2: I'd still like to see some comparison with the state-of-the-art alternatives, e.g. `niv`. `niv` does have explicit listing of dependencies, and supports automatic updating.

Reviewer 6: Not sure what "explicit listing of dependencies" and "automated updating" mean here, can't match it with motivation section items. Is it about supplying updated nixpkgs to a flake and it's transitive dependencies?

Instead of a `flake.nix`, flakes could store their metadata in a
simpler format such as JSON or TOML. This avoids the Turing tarpit
where getting flake metadata requires the execution of an arbitrarily

> complex, possibly non-terminating program.

Reviewer 5: Please use something standard like JSON. It could be incredibly useful for tools which aren't nix to be able to inspect and potentially update the flake lock, and we clearly don't need the full flexibility of the nix language for a machine-generated file.

Reviewer 4: This isn’t talking about the lockfile. It says “flake.nix”, not “flake.lock”.

Reviewer 5: Ah right, sorry.

> complex, possibly non-terminating program.

Reviewer: Maybe a part of the flake metadata (edition, description, name if added) could live in a JSON file and the rest like outputs in `flake.nix`

Author: Yeah, that's the idea.

Reviewer: Well, inputs are also almost metadata, and there keeping things in Turing-complete Nix has some benefits (and some drawbacks, sure); as written the alternative seems a bit too binary.

(I am among those who support basic part of the metdata being in JSON, but I do support input definitions near the output definitions, in Nix)

Flakes could be implemented as an external tool on top of Nix. Indeed,
there is nothing that flakes allow you to do that couldn't previously
be done using `fetchGit`, the `--pure-eval` flag and some shell
scripting. However, implementing flake-like functionality in an
external tool would defeat the goals of this RFC. First, it probably
wouldn't lead to a standard way to structure and compose Nix projects,

> since we might well end up with numerous competing

Reviewer: I think home-manager is an example where we had exactly that, then one standard got majority?

"standards". Second, it would degrade rather than improve the Nix UX,
since users would now have to deal with Nix *and* the flake-like tool

> on top of it.

Reviewer: The tool being a part of Nix doesn't do much to remove needing to keep track of flake-specific invokations

Author: Not true, commands like `nix build` or `nix dev-shell` "just work" when invoked in a flake directory.

Reviewer: This part can be achieved by specifying that flake should have a `default.nix` (that imports `flake.nix` etc.)

# Unresolved questions
[unresolved]: #unresolved-questions

* Should flakes have arguments (like "system type")? This must be done

>   in a way that maintains hermetic evaluation and evaluation caching.

Reviewer 2: This seems very important to me. Nix and nixpkgs are shaping up to be a best-in-class solution to cross-compiling complex software. It would be a shame if flakes made this harder again.

Reviewer: Could this be implemented via inputs that are literal Nix values?

* Currently, if flake dependencies (repositories or branches) get
  deleted upstream, rebuilding the flake may fail. (This is similar to
  `fetchurl` referencing a stale URL.) We need a command to gather all
  flake dependencies and copy them somewhere else (possibly vendor

>   them into the repository of the calling flake).

Reviewer: Could existing binary cache logic be reused here? In a locked flake all inputs are checksummed, so they can be built as FODs, which could be made fetchable even from untrusted caches.

Maybe this requires an extra field in flake metadata.

Author: Yes, in principle you could use a binary cache for this.

https://github.com/NixOS/nix/issues/3253

* Maybe flake metadata should be stored in a `flake.json` or
  `flake.toml` file. This would prevent ambiguities when the Nix
  language changes in a future edition.

# Future work

> [future]: #future-work

Reviewer 4: All of the CLI, etc. work should be mentioned here.

  that it is hard to ensure reproducible evaluation of NixOS or NixOps
  configurations.

* Nix projects lack discoverability and a standard structure. For
  example, it's just convention that a repository has a `release.nix`
  for Hydra jobs and a `default.nix` for packages.

* There is no standard way to compose Nix projects. Typical ways are
  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. This RFC,
however, only describes the format and semantics of flakes; it doesn't
describe changes to the `nix` command to support flakes.

# Detailed design
[design]: #detailed-design

## Flakes

A flake is a Git repository that contains a file named `flake.nix` in
the root directory. (In the future, there might be other types of

> flakes, such as Mercurial repositories or tarballs.) `flake.nix`

Reviewer 7: Later on in the text of the RFC `tarball` and `hg` are mentioned as supported input types. So does this mean that this sentence must be changed, or is it the later part that is wrong, or is it only supported for non-flake inputs?

  allowed value is `201911`. Nix rejects flakes with an unsupported
  edition.

* `description`: A short description of the flake.

* `inputs`: An attrset specifying the dependencies of the flake
  (described below).

* `outputs`: A function that, given an attribute set containing the
  outputs of each of the input flakes keyed by their identifier,
  yields the Nix values provided by this flake. Thus, in the example
  above, `inputs.nixpkgs` contains the result of the call to the
  `outputs` function of the `nixpkgs` flake, and
  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
  attribute of `nixpkgs`.

  In addition to the outputs of each input, each input in `inputs`
  also contains some metadata about the inputs. These are:

  * `outPath`: The path in the Nix store of the flake's source
    tree. This means that you could import Nixpkgs in a more
    legacy-ish way by writing

        with import inputs.nixpkgs { system = "x86_64-linux"; };

    since `nixpkgs` still contains a `/default.nix`. In this case we
    bypass its outputs entirely and only use the flake mechanism to
    get its source tree.

>   * `rev`: The commit hash of the flake's Git repository.

Reviewer 8: nitpicking: how about Mercurial changeset IDs, etc? :p

* Access to files outside of the top-level flake or its inputs, as
  well as paths fetched using `fetchTarball`, `fetchGit` and so on
  without a commit hash or content hash. In particular this means that
  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.

* Access to the environment. This means that `builtins.getEnv "<var>"`
  always returns an empty string.

* Access to the system type (`builtins.currentSystem`).

* Access to the current time (`builtins.currentTime`).

* Use of the Nix search path (`<...>`); composition must be done
  through flake inputs or `fetchX` builtins.

# Drawbacks
[drawbacks]: #drawbacks

Pure evaluation breaks certain workflows. In particular, it breaks the
use of the Nixpkgs configuration file. Similarly, there are people who
rely on `$NIX_PATH` to pass configuration data to NixOps
configurations.

# Alternatives
[alternatives]: #alternatives

For composition of multi-repository projects, the main alternative is
to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
in other repositories. However, since there is no explicit listing of
dependencies, this does not provide automatic updating (i.e. there is

> no equivalent of `nix flake update`).

Reviewer 7: `nix flake update` is only mentioned here in the new version of the RFC.

depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.x86_64-linux.dwarffs =
      with nixpkgs.packages.x86_64-linux;
      with nixpkgs.builders;
      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModified}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

    nixosModules.dwarffs = ...;

    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;

    checks.build = packages.x86_64-linux.dwarffs;
  };
}

> ```

Reviewer 9: Am I missing something obvious, or is the `inputs` attribute missing from this example?

Author: It's implicit. If the `outputs` function has an argument named `X`, and there is no `inputs.<X>` attribute, than `inputs.<X>` defaults to `{ type = "indirect"; id = "<X>"; }`, i.e. a lookup in the flake registry,

  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. This RFC,
however, only describes the format and semantics of flakes; it doesn't
describe changes to the `nix` command to support flakes.

# Detailed design
[design]: #detailed-design

## Flakes

A flake is a directory that contains a file named `flake.nix` in the
root directory. `flake.nix` specifies some metadata about the flake
such as dependencies (called *inputs*), as well as its *outputs* (the
Nix values such as packages or NixOS modules provided by the flake).

As an example, below is the `flake.nix` of
[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
for automatically fetching DWARF debug symbols by ELF build ID). It
depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.x86_64-linux.dwarffs =
      with nixpkgs.packages.x86_64-linux;

>       with nixpkgs.builders;

Reviewer 10: Just tried this example and it failed with:
```
error: attribute 'builders' missing, at /nix/store/bwa7j11q87qcs02qh9z0wdqc4sc25q4g-source/flake.nix:7:12
```

After that I ran into this issue:
```
error: attribute 'packages' missing, at /nix/store/zi21mr9l9gn20q1an9z455s0fn9zpm54-source/flake.nix:6:12
```

I ended up with the following flake file:
```nix
{
  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.x86_64-linux.dwarffs =
      with nixpkgs.packages.x86_64-linux;
#      with nixpkgs.builders;
      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModifiedDate}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

#    nixosModules.dwarffs = ...;

    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;

    checks.build = packages.x86_64-linux.dwarffs;
  };
}
```

Steps to reproduce:

1. Spawn a nix-shell with `nixFlakes`: `nix-shell -p nixFlakes` (on recent/current nixos-unstable channel, commit 029a5de0839 in my case)
2. Enable the `flakes` and `nix-command` feature in ~/.config/nix/nix.conf (verify via `nix show-config | grep experi`)
3. Write the above `flake.nix` into a new (empty?) directory
4. change to that new directory and execute `nix build flake:.`

Maybe it is worth exchanging the example with something simpler. How about a variant of the example(s) presented in the blog post?

      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModifiedDate}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

>     nixosModules.dwarffs = ...;

Reviewer 11: Can I nest modules here? i.e. lenovo.thinkpad.x280
https://github.com/NixOS/nixos-hardware/pull/176#discussion_r450251516

Author: No, the current flake outputs (except `legacyPackages`) don't support nested attrsets in order to make it possible to query them efficiently.

Reviewer 12: I noticed this as well and wonder how we are going to handle sub package sets then. Maybe that's something for later after this RFC is accepted.

Reviewer 11: We definitely need python3Packages.<foo> to work.

    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;

    checks.build = packages.x86_64-linux.dwarffs;
  };
}
```

> A flake has the following attributes:

Reviewer 10: Are those the only allowed values? Is a `flake.nix` invalid if it has other attribute?

Author: Yes, any other attribute is an error.

Reviewer 10: Perhaps we should add that to the document. It currently leaves space for interpretation.

Reviewer 13: @edolstra would it then make sense to have a top-level attribute like `meta` for arbitrary configuration one might want per flake?

input. The special input named `self` refers to the outputs and source
tree of *this* flake. Thus, a typical `outputs` function looks like
this:

    outputs = { self, nixpkgs, import-cargo }: {
      ... outputs ...
    };

It is also possible to omit inputs entirely and *only* list them as
expected function arguments in `outputs`. Thus,

    outputs = { self, nixpkgs }: ...;

without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in
the flake registry.

Repositories that don't contain a `flake.nix` can also be used as
inputs, by setting the input's `flake` attribute to `false`:

    inputs.grcov = {
      type = "github";
      owner = "mozilla";
      repo = "grcov";
      flake = false;
    };

    outputs = { self, nixpkgs, grcov }: {
      packages.x86_64-linux.grcov = stdenv.mkDerivation {
        src = grcov;
        ...
      };
    };

The following input types are specified at present:

* `git`: A Git repository or dirty local working tree.

* `github`: A more efficient scheme to fetch repositories from GitHub
  as tarballs. These have slightly different semantics from `git`
  (in particular, the `revCount` attribute is not available).

> * `gitlab`: Like `github`, but for Git repositories hosted on GitLab.

Reviewer 8: I don't think this should have been dropped in during the FCP, as it seems very underspecified. It doesn't clearly define the parameters that this input takes. GitLab can be self-hosted (so you can't assume the repo is at `gitlab.com`, is there a way to change that?), \*and\* it has a concept of (recursively nested) subgroups, which don't really play well with the `owner`/`repo` keys that `github` uses.

* `tarball`: A `.tar.{gz,xz,bz2}` file.

* `path`: A directory in the file system. This generally should be
  avoided in favor of `git` inputs, since `path` inputs have no
  concept of revisions (only a content hash) or tracked files
  (anything in the source directory is copied).

* `hg`: A Mercurial repository.

Transitive inputs can be overriden from a `flake.nix` file. For
example, the following overrides the `nixpkgs` input of the `nixops`
input:

    inputs.nixops.inputs.nixpkgs = {
      type = "github";
      owner = "my-org";
      repo = "nixpkgs";
    };

> It is also possible to "inherit" an input from another input. This is

Reviewer 14: I have a usability concern here:

If I understand it correctly, the default behaviour is to use the lock file of each flake to resolve its inputs. This would potentially lead to many different nixpkgs versions being used, creating long build times and large output closures with many duplicate files. It would also enable old versions of libraries to sneak in, creating a potential security concern for users who are not very familiar with Nix and use the default behaviour.
Being able to reproduce the same result is great, but if the same version is always reproduced by default, what is the cost for that?

Or is the lock file of the top level flake used to provide the inputs for all dependency flakes recursively?

Author: Yes, there can be many versions of nixpkgs in the flake input closure. However, if you're using Nixpkgs overlays or NixOS modules, only the top-level nixpkgs version matters, since overlays/modules get composed with the top-level nixpkgs. If needed you can override transitive inputs to point to the same version, e.g.
```
  inputs.nixpkgs.url = "nixpkgs/nixos-20.03";
  inputs.dwarffs.inputs.nixpkgs.follows = "nixpkgs";
  inputs.nix.inputs.nixpkgs.follows = "nixpkgs";
  inputs.blender-bin.inputs.nixpkgs.follows = "nixpkgs";
```

Reviewer 14: This answers my question. If everyone uses overlays, this will not be an issue. It will be important to communicate this as the preferred way of providing package outputs in documentation. I will raise concerns again if I think this is not sufficiently communicated.

Reviewer 11: I see a usability problem with composing independent overlays. This can easily lead to hard to debug recursion errors or conflicts if they override the same packages in different ways. For that reason I decided to not use overlays in NUR but having namespaces not conflicting with each other.

Reviewer 11: > Yes, there can be many versions of nixpkgs in the flake input closure. However, if you're using Nixpkgs overlays or NixOS modules, only the top-level nixpkgs version matters, since overlays/modules get composed with the top-level nixpkgs. If needed you can override transitive inputs to point to the same version, e.g.
> 
> ```
>   inputs.nixpkgs.url = "nixpkgs/nixos-20.03";
>   inputs.dwarffs.inputs.nixpkgs.follows = "nixpkgs";
>   inputs.nix.inputs.nixpkgs.follows = "nixpkgs";
>   inputs.blender-bin.inputs.nixpkgs.follows = "nixpkgs";
> ```

How does this work if `dwarffs` has a dependency flake that depends on nixpkgs?
Like this?

```
   inputs.dwarffs.dependency.inputs.nixpkgs.follows = "nixpkgs";
```

Or is inputs flattened?

Author: > Note that even if only modules/overlays are used in the current implementation it downloads
nixpkgs for those flakes anyways also its are never used

No, flake inputs are fetched lazily, so if you don't use the `nixpkgs` input of the `dwarffs` input, it won't be fetched.

there is nothing that flakes allow you to do that couldn't previously
be done using `fetchGit`, the `--pure-eval` flag and some shell
scripting. However, implementing flake-like functionality in an
external tool would defeat the goals of this RFC. First, it probably
wouldn't lead to a standard way to structure and compose Nix projects,
since we might well end up with numerous competing
"standards". Second, it would degrade rather than improve the Nix UX,
since users would now have to deal with Nix *and* the flake-like tool
on top of it.

# Unresolved questions
[unresolved]: #unresolved-questions

* Should flakes have arguments (like "system type")? This must be done
  in a way that maintains hermetic evaluation and evaluation caching.

* Currently, if flake dependencies (repositories or branches) get
  deleted upstream, rebuilding the flake may fail. (This is similar to
  `fetchurl` referencing a stale URL.) We need a command to gather all
  flake dependencies and copy them somewhere else (possibly vendor
  them into the repository of the calling flake).

* Maybe flake metadata should be stored in a `flake.json` or
  `flake.toml` file. This would prevent ambiguities when the Nix

>   language changes in a future edition.

Reviewer 15: Sorry for being so late to the party, but while I understand that there were valid reasons to remove the `edition`-field in the end, wouldn't it make sense to keep a similar flag to make sure that `flake.*` files are interpreted properly even if we switch back and forth between e.g. `.nix`, `.toml` and sth else?

I'm asking because @edolstra stated in https://github.com/NixOS/rfcs/pull/49#issuecomment-637640007 that it's not 100% clear whether `.nix`-files are the way to go for flakes and I could imagine that we'd run into problems if we switch back to `.nix` with a different structure after several iterations.

Reviewer 13: The decision was to name next version `flakes2.nix` or similar so there wouldn't be complex need of interpreting such files in a different manner.

  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
  environment variables, and Git repositories. This means for instance
  that it is hard to ensure reproducible evaluation of NixOS or NixOps
  configurations.

* Nix projects lack discoverability and a standard structure. For
  example, it's just convention that a repository has a `release.nix`
  for Hydra jobs and a `default.nix` for packages.

* There is no standard way to compose Nix projects. Typical ways are
  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. This RFC,
however, only describes the format and semantics of flakes; it doesn't
describe changes to the `nix` command to support flakes.

# Detailed design
[design]: #detailed-design

## Flakes

> A flake is a directory that contains a file named `flake.nix` in the

Reviewer 10: According to the current [implementation](https://github.com/NixOS/nix/blob/8c75621da6260d92459e105892751457e1659741/src/libexpr/flake/flakeref.cc#L126) a directory is only a flake if it is in a git repository:
```
error: path '/tmp/test-flake' is not a flake (because it does not reference a Git repository)
```
I can't see the wording here that requires using Git.

Is this a bug in the RFC or in the implementation?

Reviewer 16: *ping* I'd be interested in this as well.

Author: This is no longer the case. You can also have a plain directory containing a `flake.nix` file.

depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.x86_64-linux.dwarffs =
      with nixpkgs.packages.x86_64-linux;
      with nixpkgs.builders;
      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModifiedDate}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

    nixosModules.dwarffs = ...;

    defaultPackage.x86_64-linux = packages.x86_64-linux.dwarffs;

>     checks.build = packages.x86_64-linux.dwarffs;

Reviewer 16: The `checks` attribute doesn't have any special meaning or does it?

Reviewer 17: If I understand correctly `nix flake check` will check the semantics of `flake.nix` and also run this checks.

Reviewer 16: I guess if it has a special purpose this should be documented here?

Author: This RFC doesn't specify any particular flake outputs since that's up to tools like the `nix` command. So their appearance here is just an example.

  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
  attribute of `nixpkgs`.

  In addition to the outputs of each input, each input in `inputs`
  also contains some metadata about the inputs. These are:

  * `outPath`: The path in the Nix store of the flake's source
    tree. This means that you could import Nixpkgs in a more
    legacy-ish way by writing

        with import inputs.nixpkgs { system = "x86_64-linux"; };

    since `nixpkgs` still contains a `/default.nix`. In this case we
    bypass its outputs entirely and only use the flake mechanism to
    get its source tree.

  * `rev`: The commit hash of the flake's repository, if applicable.

  * `revCount`: The number of ancestors of the revision `rev`. This is
    not available for `github` repositories (see below), since they're
    fetched as tarballs rather than as Git repositories.

  * `lastModifiedDate`: The commit time of the revision `rev`, in the
    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
    this is available for both Git and GitHub repositories, so it's
    useful for generating (hopefully) monotonically increasing version

>     strings.

Reviewer 6: Lost here as to what and why is lastModifiedDate. Is it an alternative for rev? Or is it an alternative for revCount (which then doesn't make sense for tarball/github fetcher)? Would a option tag serve the purpose?
"hopefully" and "version string" combination sounds scary, what is it expected to be used for?

Author: `revCount` and `lastModifiedDate` are primarily there to help you generate version strings (e.g. in the `name` attribute of a derivation). For example, the `dwarffs` package from the `dwarffs` flake has a version like `0.1.20200511`. `lastModifiedDate` is indeed an alternative to `revCount` since the latter isn't always available.

The "hopefully" refers to the fact that you can always fake timestamps with Git, but that's a case of "just don't do that" :-)

  * `lastModified`: The commit time of the revision `rev` as an integer
    denoting the number of seconds since 1970.

  * `narHash`: The SHA-256 (in SRI format) of the NAR serialization of
    the flake's source tree.

  The value returned by the `outputs` function must be an attribute
  set. The attributes can have arbitrary values; however, some tools
  may require specific attributes to have a specific value (e.g. the
  `nix` command may expect the value of `packages.x86_64-linux` to be
  an attribute set of derivations built for the `x86_64-linux`
  platform).

## Flake inputs

The attribute `inputs` specifies the dependencies of a flake. These
specify the location of the dependency, or a symbolic flake identifier
that is looked up in a registry or in a command-line flag. For

> example, the following specifies a dependency on the Nixpkgs and Hydra

Reviewer 18: ```suggestion
example, the following specifies a dependency on the Nixpkgs and import-cargo
```

repositories:

    # A GitHub repository.
    inputs.import-cargo = {
      type = "github";
      owner = "edolstra";
      repo = "import-cargo";
    };

    # An indirection through the flake registry.
    inputs.nixpkgs = {
      type = "indirect";
      id = "nixpkgs";
    };

Each input is fetched, evaluated and passed to the `outputs` function
as a set of attributes with the same name as the corresponding
input. The special input named `self` refers to the outputs and source
tree of *this* flake. Thus, a typical `outputs` function looks like
this:

    outputs = { self, nixpkgs, import-cargo }: {
      ... outputs ...
    };

It is also possible to omit inputs entirely and *only* list them as
expected function arguments in `outputs`. Thus,

    outputs = { self, nixpkgs }: ...;

without an `inputs.nixpkgs` attribute will simply look up `nixpkgs` in

> the flake registry.

Reviewer 8: There is no specification for any sort of flake registry in this spec, what is this referring to?

Reviewer 19: https://github.com/NixOS/flake-registry

AFAIK it's like aliases for longer URLs

(there's a section on it in the blog post above)

Reviewer 8: Well yeah, but that's not part of the spec here, is it? I don't think it's a good idea to go ahead and specify "this implicitly talks to the flake registry" without specifying how the flake registry behaves, and how it will be ran.

Reviewer 19: True. Maybe it was dropped as part of this: https://github.com/NixOS/rfcs/pull/49/files#r354288001

Repositories that don't contain a `flake.nix` can also be used as
inputs, by setting the input's `flake` attribute to `false`:

    inputs.grcov = {
      type = "github";
      owner = "mozilla";
      repo = "grcov";
      flake = false;
    };

    outputs = { self, nixpkgs, grcov }: {
      packages.x86_64-linux.grcov = stdenv.mkDerivation {
        src = grcov;
        ...
      };
    };

The following input types are specified at present:

> * `git`: A Git repository or dirty local working tree.

Reviewer 6: git protocol doesn't let you get to the working tree, so should be two-three different options
- git repo (at tag/branch/commit)
- local path
- git repo (at tag/branch/commit) + path inside repo

Author: I'll have to check - it might be that this description is out of date. It may be that `git+file:///path` already doesn't support a dirty tree.

* `github`: A more efficient scheme to fetch repositories from GitHub
  as tarballs. These have slightly different semantics from `git`

>   (in particular, the `revCount` attribute is not available).

Reviewer 6: imo github option shouldn't be a part of the spec (but may have special shortcuts/tools in cli)
- tarball fetch is just an instance of http fetch
- the way github generates tarball urls may change in future
- for private repos user may have configured ssh access but not http (via tokens)

Reviewer 20: It would be nice for things like github to be provided by external plugins, but github fetch *does* have different semantics than http fetch, in part because of the things you call out here! IMO it should be a separate primitive.

Reviewer 6: Bitbucket for example also has tarball downloads of git repos, GitLab too, likely many others. So going this way would probably mean that in future more git (or even non-git) hosting services will need special handling.

Reviewer 21: Other hosting services aren't the center of NixOS coordination, though..

Author: Right, most Nix-related repos are on GitHub, so in the interest of UX, it's nice to provide a shorter syntax and better semantics for GitHub. We don't have to support every other hosting service out there.

Reviewer 22: I find that most repos use `feature/<name>` branches, but I didn't find a way to make a URI containing this branch name. Is this expected behavior?

Reviewer 22: > Right, most Nix-related repos are on GitHub, so in the interest of UX, it's nice to provide a shorter syntax and better semantics for GitHub. We don't have to support every other hosting service out there.

@rycee uses many GitLab inputs for the `home-manager` repo. Is GitLab may be also desired?

Author: Yes, especially if we can get the `rev` attribute from GitLab.

Reviewer 22: I create a Libfetcher for GitLab: https://github.com/NixOS/nix/pull/3636

  build this flake, the input `nixpkgs` is mapped to revision
  `7f8d4b088e2df7fdb6b513bc2d6941f1d422a013` of the `edolstra/nixpkgs`
  repository on GitHub.

* `flake`: A Boolean denoting whether this is a flake or non-flake
  dependency. Corresponds to the `flake` attribute in the `inputs`
  attribute in `flake.nix`.

The `info`, `original` and `locked` attributes are omitted for the
root node. This is because we cannot record the commit hash or content
hash of the root flake, since modifying `flake.lock` will invalidate
these.

The graph representation of lock files allows circular dependencies
between flakes. For example, here are two flakes that reference each
other:
```
{
  inputs.b = ... location of flake B ...;
  # Tell the 'b' flake not to fetch 'a' again, to ensure its 'a' is
  # *this* 'a'.
  inputs.b.inputs.a.follows = "";
  outputs = { self, b }: {
    foo = 123 + b.bar;
    xyzzy = 1000;
  };
}
```
and
```
{
  inputs.a = ... location of flake A ...;
  inputs.a.inputs.b.follows = "";
  outputs = { self, a }: {
    bar = 456 + a.xyzzy;
  };
}
```

Lock files transitively lock direct as well as indirect

> dependencies. That is, if a lock file exists and is up to date, Nix

Reviewer 6: What does it mean for lock file to be up to date?

Author: "Up to date" == "all inputs have a corresponding entry in the lock file". If there is a lock, then it won't be updated unless you pass a flag like `--update-input`.

---
feature: flakes
start-date: 2019-07-09
author: Eelco Dolstra
co-authors: TBD
shepherd-team: Domen Kožar, Alyssa Ross, Shea Levy, John Ericson
shepherd-leader: Domen Kožar
related-issues: (will contain links to implementation PRs)
---

# Summary
[summary]: #summary

This RFC proposes a mechanism to package Nix expressions into
composable entities called "flakes". Flakes allow hermetic,
reproducible evaluation of multi-repository Nix projects; impose a
discoverable, standard structure on Nix projects; and replace previous
mechanisms such as Nix channels and the Nix search path.

# TLDR

* A prototype implementation is available in Nixpkgs: `nix run
  nixpkgs.nixFlakes`.

> * Flakes replace channels. For example,

Reviewer: Are they supposed to also fit some of the overlay use cases?

Author: Not really. They can be a delivery mechanism for overlays, though.

  depends on the `nixpkgs` flake. To ensure reproducibility,
  dependencies are pinned to specific versions using a *lock file*
  (e.g. [for
  patchelf](https://github.com/NixOS/patchelf/blob/master/flake.lock)). Lock
  files are generated automatically.

* Flakes are evaluated in pure mode, meaning they can't access
  anything other than their own source or declared dependencies. This
  allows `nix` to cache evaluation results.

* The `nix` command is flake-based. For example,
  ```
  # nix build
  ```
  builds the flake in the current directory.

# Motivation
[motivation]: #motivation

Flakes are motivated by a number of serious shortcomings in Nix:

* While Nix pioneered reproducible builds, sadly, Nix expressions are
  not nearly as reproducible as Nix builds. Nix expressions can access
  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
  environment variables, and Git repositories. This means for instance
  that it is hard to ensure reproducible evaluation of NixOS or NixOps
  configurations.

* Nix projects lack discoverability and a standard structure. For
  example, it's just convention that a repository has a `release.nix`

>   for Hydra jobs and a `default.nix` for packages.

Reviewer: Isn't `default.nix` and `shell.nix` assumed by enough tools to be a bit more than mere convention?

Author: No, `default.nix` has no well-known structure. I've seen it contain a single package at top-level, an attrset of packages, and an attrset of arbitrary stuff (like functions).

* There is no standard way to compose Nix projects. Typical ways are
  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. It can be
seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
rather than a separate tool.

# Detailed design
[design]: #detailed-design

## Flakes

> A flake is a Git repository that contains a file named `flake.nix` in

Reviewer 23: Defining flakes explicitly as git repos seems like a useful stepping-stone toward a delightful future but I sense that it may even now be too restrictive for what I think even early-majority flake adopters will want, let alone what the long tail of users will want. 

As a result, would it be possible to sketch out at least a little bit, here or elsewhere, where the current design is driven to depend exclusively on git for short-term convenience, expediency, or concreteness vs more long-term reasons so that we can think together about how the needs or aspirations those reasons represent might be addressed in non-git scenarios e.g., involving different network, security, compatibility, usability, or scale considerations?

Reviewer 23: To add a couple of high-level use cases motivating the question beyond more abstract design value judgments about configurability and coupling:

* while acknowledging the importance of human-meaningful flake metadata for good UX, as a user experimenting with flakes for the first time, I would like to be able to define flakes in local directories on my filesystem that are not yet version-controlled or that are version-controlled with something other than git

* as a user who occasionally interacts with monorepos, some of them large, some of them not based on git, or some of them based on git but using it in non-conventional ways, I would like flakes to be less tightly coupled not only with git but also with the idea that relevant units of code live *precisely* in "git repositories made available by a network service via URLs"

* finally, as a user interested in using nix flakes in diverse environments including on hardware or operating systems where git may not be available, as well as with colleagues who are not yet familiar with git or where software is being managed by other means, I would like the flakes design to be a little bit more agnostic about how flakes and their supporting resources are organized, aggregated, and transported even if awesome support for git and GitHub remain top-priority concerns.

Author: There is nothing about flakes that makes them inherently tied to Git or GitHub. It's just that those are the only two mechanisms implemented at the moment. Mercurial and tarball support would be straightforward to add.

  * `revCount`: The number of ancestors of the revision `rev`. This is
    not available for `github:...` repositories (see below), since
    they're fetched as tarballs rather than as Git repositories.

  * `lastModified`: The commit time of the revision `rev`, in the
    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
    this is available for both Git and GitHub repositories, so it's
    useful for generating (hopefully) monotonically increasing version
    strings.

  The special input named `self` refers to the outputs and source tree
  of *this* flake.

  A number of outputs have a special meaning, as discussed below.

## Well-known outputs

A number of outputs have a specific meaning to Nix or other tools like
Hydra. Currently, these are:

* `packages`: A set of derivations used as a default by most `nix`
  commands. For example, `nix run nixpkgs:hello` uses the
  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain
  any non-derivation attributes. This also means it cannot be a nested
  set! (The rationale is that supporting nested sets requires Nix to
  evaluate each attribute in the set, just to discover which packages
  are provided.)

> * `defaultPackage`: A derivation used as a default by most `nix`

Reviewer: What happens if `packages.defaultPackage` also exists? Check complains?

Author: No, but that would be a good idea to check.

Reviewer: Maybe it is a good idea to specify that it should not exist then?

  with the constraint that `<path>` must end with `.git` for
  non-`file` repositories. `<params>` are a list of key/value pairs in
  URI query parameter syntax. The following parameters are supported:

  * `ref`: The branch or tag to fetch. The default is `master`.
  * `rev`: The Git commit hash to fetch. Note that this commit must be
    an ancestor of `ref`, since Nix doesn't clone the entire
    repository, only the specified `ref` (and the Git protocol doesn't
    allow fetching a `rev` without a known `ref`). The default is the
    commit currently pointed to by `ref`.
  * `dir`: The subdirectory of the repository in which `flake.nix` is
    located. This parameter enables having multiple flakes in a
    repository. The default is the root directory.

  For example, the following are valid Git flake references:

  * `https://example.org/my/repo.git`
  * `https://example.org/my/repo.git?dir=flake1`
  * `ssh://git@github.com:NixOS/nix.git?ref=v1.2.3`
  * `git://github.com/edolstra/dwarffs.git?ref=unstable&rev=e486d8d40e626a20e06d792db8cc5ac5aba9a5b4`
  * `file:///home/my-user/some-repo/some-repo.git`

* Local paths. These have the form

      <path>(\?<params)?

>   where `<path>` must refer to (a subdirectory of) a Git repository.

Reviewer: Can we please support plain directories? Then we don't need to wait until at least Mercurial is supported, not to mention negotiating about Fossil, Pijul, Monotone…

Reviewer 4: I think this at least is worth thinking about before we make any sort of flakes release, because I think otherwise it could cause compatibility problems down the line, especially in regards to:

>    * The `dir` parameter is automatically derived. For example, if `/foo/bar` is a Git repository, then the flake reference `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.

How would we resolve this? Would such a directory be a git flake with a `dir` parameter, or a `dir` repository. What if you wanted the other? Would there be a way to override it?

Reviewer 4: I think it’s also important that it would at least be possible to implement fetching remote tarballs, even if we don’t support it right away. I don’t see any immediate compatibility concerns with that, though.

Reviewer 8: requiring local paths to be git repos would be annoying (how would this behave when nesting git repositories?) and there's nothing stopping me from having a `?` inside a path; not being able to tab-complete those would be a pain..

Reviewer 8: not only that, but my current workflow (nixpkgs with custom patches inside a git subtree) is heavily discouraged, as this means i have a `flakes.nix` not in the root of a git repo, which the spec only barely seems to support.

  These differ from `file://` flake references in a few ways:

  * They refer to the working tree (unless an explicit `rev` or `ref`
    is specified), so evaluation can access dirty files. (Dirty files
    are files that are tracked by git but have uncommitted changes.)

  * The `dir` parameter is automatically derived. For example, if
    `/foo/bar` is a Git repository, then the flake reference
    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.

* GitHub repositories. These are downloaded as tarball archives,
  rather than through Git. This is often much faster and uses less
  disk space since it doesn't require fetching the entire history of
  the repository. On the other hand, it doesn't allow incremental
  fetching (but full downloads are often faster than incremental
  fetches!). The syntax is:

      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?

  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a

>   commit hash (`rev`). Note that unlike GitHub, Git allows fetching by

Reviewer 24: ```suggestion
  commit hash (`rev`). Note that unlike Git, GitHub allows fetching by
```

Reviewer 25: Addressed in tweag/rfcs#1

evaluation. It is also necessary to prevent certain impurities. In
particular, the `nix` command now defaults to evaluating in "pure"
mode, which means that the following are disallowed:

* Access to files outside of the top-level flake or its inputs, as
  well as paths fetched using `fetchTarball`, `fetchGit` and so on
  without a commit hash or content hash. In particular this means that
  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.

* Access to the environment. This means that `builtins.getEnv "<var>"`
  always returns an empty string.

* Access to the system type (`builtins.currentSystem`).

* Access to the current time (`builtins.currentTime`).

* Use of the Nix search path (`<...>`); composition must be done
  through flake inputs or `fetchX` builtins.

Pure mode can be disabled by passing `--impure` on the command line.

## Evaluation caching

The fact that by default (i.e. in pure mode) evaluation is hermetic
allows `nix` to cache flake attributes. For example (doing `nix build`
on an already present package):

    $ time nix build nixpkgs:firefox
    real    0m1.497s

    $ time nix build nixpkgs:firefox
    real    0m0.052s

> The evaluation cache is kept in `~/.cache/nix/eval-cache-v1.sqlite`,

Reviewer 26: Given that nix already has a mechanism for storing values identified by the hash of their inputs (i.e. /nix/store), it would seem conceptually neater to reuse that. It would also allow sharing some other utilities down the line (e.g. garbage collection of old entries) instead of reimplementing them.

It's possible that is impractical for performance reasons, but then maybe we can somehow speed up /nix/store for trivially small entries instead? Or put these in /nix/store as the source of truth, as well as in a sqlite cache on top of that.

Author: Using the Nix store would be orders of magnitude slower and take up way more disk space, so I don't think it's an option.

rely on `$NIX_PATH` to pass configuration data to NixOps
configurations.

# Alternatives
[alternatives]: #alternatives

For composition of multi-repository projects, the main alternative is
to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
in other repositories. However, since there is no explicit listing of
dependencies, this does not provide automatic updating (i.e. there is
no equivalent of `nix flake update`).

Instead of a `flake.nix`, flakes could store their metadata in a
simpler format such as JSON or TOML. This avoids the Turing tarpit
where getting flake metadata requires the execution of an arbitrarily
complex, possibly non-terminating program.

# Unresolved questions
[unresolved]: #unresolved-questions

* How to handle the system type? Currently `x86_64-linux` is
  hard-coded everywhere.

* How to do Nixpkgs overlays? In principle, overlays would just be a
  kind of flake output.

* More in general, how to handle flake arguments? This must be done in
  a way that maintains hermetic evaluation and evaluation caching.

* What are the criteria for inclusion in the global flake registry?

> * Hammer out the details of NixOS/NixOps support for flakes.

Reviewer 13: That would rock, as it's the main use of channels.

* While Nix pioneered reproducible builds, sadly, Nix expressions are
  not nearly as reproducible as Nix builds. Nix expressions can access
  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
  environment variables, and Git repositories. This means for instance
  that it is hard to ensure reproducible evaluation of NixOS or NixOps
  configurations.

* Nix projects lack discoverability and a standard structure. For
  example, it's just convention that a repository has a `release.nix`
  for Hydra jobs and a `default.nix` for packages.

* There is no standard way to compose Nix projects. Typical ways are
  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems. It can be
seen as a "Cargo/NPM/... for Nix" except that it's built into Nix

> rather than a separate tool.

Reviewer 7: Let me insist on something because I read some really odd things while going back in the discussion:

>thinking about making nix a package manager

Nix is *already* a package manager. Just because it doesn't conform with what you are used to with *some* package managers like npm and Cargo does not make it less a package manager. Package managers have an almost 30-year long history, with very varied design decisions. And even today, npm and Cargo (and RubyGems and PyPI) do not represent the only way to make a good modern package manager.

> rather than a separate tool.

Reviewer 25: Isn't the point of this comparison that flakes would make it easier to package language-specific package manager ecosystems for nix-the-non-language-specific-package-manager?

As an example, currently there's a couple GHCs in nixpkgs, if you need one that isn't you're out of luck. Flakes would decouple the GHCs needed by nixpkgs from the ones that can be made available. Flakes'd be more like stackage LTSes, available for as long as anyone needs them. (Note that I'm not saying it's impossible to pin nixpkgs to the latest revision that had a GHC or that other technical solutions exist. I am saying flakes'd turn this from a PITA into simply specifying the right flake.)

> rather than a separate tool.

Reviewer 7: Sure, it would make such a use simpler and this is one of the great things about flakes.

But this sentence as is can be interpreted in many ways. If everyone who reads it interprets it differently, it's better to just remove it. This RFC is overall very clear, but this sentence does not help.

> rather than a separate tool.

Reviewer 27: Perhaps it'd be better to rephrase as "Cargo/NPM/... for the [Nix Expression Language](https://nixos.wiki/wiki/Nix_Expression_Language)"?

Otherwise as @Zimmi48 mentions this is conflated with Nix The Package Manager(this is something that also got me confused).

Also, I wonder if there is some shorthand the community uses to refer to the Nix Expression Language(NixEx?).

> rather than a separate tool.

Reviewer 25: @steve-chavez, nix-lang is the name of the irc channel so I'd go with that.

As your understanding of the paragraph is almost completely orthogonal to mine (In my understanding flakes only happen to be able to provide a way to distribute nix libraries kind of by accident, their main merit is to modularize package repositories.) I now strongly agree it needs rephrasing.

> rather than a separate tool.

Reviewer 4: > Perhaps it'd be better to rephrase as "Cargo/NPM/... for the [Nix Expression Language](https://nixos.wiki/wiki/Nix_Expression_Language)"?
> 
> Otherwise as @Zimmi48 mentions this is conflated with [Nix The Package Manager](https://nixos.wiki/wiki/Nix_Package_Manager)(this is something that also got me confused).

If we can’t even describe what it is we want to create except by using the names of other things, that should give us pause. We can’t say on the website “Flakes are npm for Nix” and expect to be taken seriously.

> rather than a separate tool.

Reviewer 25: That doesn't mean similes can't be useful.

> rather than a separate tool.

Reviewer 4: That’s true, but we still haven’t come up with a better generic term for it than “package manager”.

> rather than a separate tool.

Reviewer 20: @toonn I definitely meant "package manager for nix-lang" in my initial proposal that got turned into flakes.

@alyssais I don't think the fact of naming difficulty here implies what you're thinking. The fact is, we have a programming language whose primary domain is managing packages, it's unsurprising that we have to be a bit verbose/analogous when trying to refer to the role of the language applied to the language itself.

> rather than a separate tool.

Reviewer 27: The RFC could be more clear that this is about `nix-lang`. Some parts also mention `Nix expressions`, which is not elucidating. Perhaps also rename those to `nix-lang expressions`? .

> rather than a separate tool.

Reviewer 25: Imo, Nix expression is probably the clearest way of describing this. It unambiguously refers to an expression written in the Nix Expression Language.

> rather than a separate tool.

Reviewer 28: I do think that if we had more names / "brands", it would make the layering concerns I commented on easier to resolve.

> rather than a separate tool.

Author: I've removed this sentence.

* Registry entries specified on the command line.

A registry is a JSON file that looks like this:

```
{
    "version": 1,
    "flakes": {
        "nixpkgs": {
            "uri": "github:NixOS/nixpkgs"
        },
        ...
    }
}
```

With this registry, flake references resolve as follows:

* `nixpkgs` -> `github:NixOS/nixpkgs`
* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`

The registries are searched in reverse order. Thus the local registry
overrides the global registry, and the command line takes precedence
over the local and global registries.

## `nix` command line interface

I propose to make flakes the primary way to specify packages in the
`nix` command line interface. Note that the `nix` UI is still marked
as experimental, so we have some freedom to make incompatible
changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
`nix-instantiate`) should not be changed to avoid breakage.

Most `nix` subcommands work on a list of arguments called
"installables" for lack of a better word. For example,

>     # nix run nixpkgs#hello dwarffs#dwarffs

Reviewer 29: The hash mark here seems bad! Can't we use something like ```.``` that is already in Nix syntax? Otherwise we end up making everyone quote all their command line args, due to shell using ```#``` as a comment syntax.

Reviewer 30: No need to quote. It needs to be separated by a space first

```
~ $ echo hi#there
hi#there
```

Spaces in URLs will be encoded ( `%20`)

Reviewer: Well, in ZSH quoting is still needed.

Reviewer 30: ```
clancywiggum% echo $ZSH_VERSION
5.7.1
clancywiggum% echo hi#there
hi#there
clancywiggum% echo it#does#not
it#does#not

```

Author: The separator used to be a colon (`:`) until a week ago or so, when I changed it a hash because it's less ambiguous and fits in a bit nicer with the URL-like syntax. The reason why I didn't use the hash from the start was that like Matthew I assumed incorrectly that it would be interpreted as a comment.

takes two flake-based installables. The general form is:

    <flake-ref>(#<attr-path>)?

Examples of installables:

* `nixpkgs#packages.hello`
* `nixpkgs#hello` - short for `nixpkgs:packages.hello`
* `nixpkgs/release-19.03#hello` - overrides the Git branch to use
* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082#hello` -
  specifies the exact Git revision to use
* `dwarffs` - short for `dwarffs#defaultPackage`
* `nix#hydraJobs.build.x86_64-darwin`

> * `.#hydraJobs.build.x86_64-darwin` - refers to the flake in the

Reviewer 29: ```.#``` here seems especially ugly. Some users may see examples like this and end up thinking this is a comment in your shell script.

Author: We could allow the dot to be left out, but then the hash would need to be quoted so it doesn't save any typing.

in dependencies (if they exist) are ignored.

When you build a local repository (e.g. `nix build /path/to/repo`),
Nix automatically creates a `flake.lock` file if it doesn't already
exists. It will also update the lock file if inputs are added or
removed. You can pass `--recreate-lock-file` to force Nix to recreate
the lock file from scratch (and thus check for the latest version of
each input).

Lock files are not sufficient by themselves to ensure reproducible
evaluation. It is also necessary to prevent certain impurities. In
particular, the `nix` command now defaults to evaluating in "pure"
mode, which means that the following are disallowed:

* Access to files outside of the top-level flake or its inputs, as
  well as paths fetched using `fetchTarball`, `fetchGit` and so on
  without a commit hash or content hash. In particular this means that
  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.

* Access to the environment. This means that `builtins.getEnv "<var>"`
  always returns an empty string.

* Access to the system type (`builtins.currentSystem`).

* Access to the current time (`builtins.currentTime`).

> * Use of the Nix search path (`<...>`); composition must be done

Reviewer 29: What rationale is there for continuing the ```<...>``` support in pure mode? It seems like it is a commonly disliked feature of Nix that we have a chance to remove! I would rather this be disallowed completely and instead rely on ordinary function arguments in these cases.

Author: Maybe you misread that, because `<...>` *isn't* allowed in pure mode. (That was already the case before flakes.)

>   through flake inputs or `fetchX` builtins.

Reviewer 29: Does this mean that the ```fetchX``` builtins can still be used impurely? For instance can I do this:

```
(builtins.fetchTarball <nixpkgs>).test
```

when I haven't added nixpkgs to my flakes.nix?

Author: No, they can't. You can only use the `fetchX` builtins if you supply a content/commit hash. (There's something to be said for banning them completely in flakes. That way, all external dependencies must be specified in `inputs`. But that might be going a bit too far right now.)

Reviewer: Given that we have more single-project fetchers in Nixpkgs than will be ever supported inside Nix, allowing use of FOD for import does sound like a good idea…

The flakes mechanism seeks to address all these problems. It can be
seen as a "Cargo/NPM/... for Nix" except that it's built into Nix
rather than a separate tool.

# Detailed design
[design]: #detailed-design

## Flakes

A flake is a Git repository that contains a file named `flake.nix` in
the root directory. (In the future, there might be other types of
flakes, such as Mercurial repositories or tarballs.) `flake.nix`
specifies some metadata about the flake such as dependencies (called
*inputs*), as well as its *outputs* (the Nix values such as packages
or NixOS modules provided by the flake).

As an example, below is the `flake.nix` of
[`dwarffs`](https://github.com/edolstra/dwarffs) (a FUSE filesystem
for automatically fetching DWARF debug symbols by ELF build ID). It
depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  name = "dwarffs";

  edition = 201906;

  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  inputs = [ "nixpkgs" ];

  outputs = inputs: rec {
    packages.dwarffs =
      with inputs.nixpkgs.packages;
      with inputs.nixpkgs.builders;

>       with inputs.nixpkgs.lib;

Author: Hm, what do you mean with aliased?

>       with inputs.nixpkgs.lib;

Reviewer 31: Just something like:
```
let
  pkgs = inputs.nixpkgs.packages;
  inherit (inputs.nixpkgs) builders lib;
```

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 inputs.self.lastModified}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = inputs.self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

    nixosModules.dwarffs = import ./module.nix inputs;

    defaultPackage = packages.dwarffs;

    checks.build = packages.dwarffs;
  };
}
```

A flake has the following attributes:

> * `edition`: A number that specifies the version of the flake

Reviewer 13: (nitpick) while `edition` is a valid term, it's somewhat vague to me. It's common to use `revision` or `version`.

Reviewer 31: I think this is inspired by https://doc.rust-lang.org/nightly/edition-guide/editions/index.html

Reviewer 13: But it has a very different meaning. In rust edition is opt-in: `When a new edition becomes available in the compiler, crates must explicitly opt in to it to take full advantage`.

According to text below it seems Nix will force the user to upgrade the version, so it's a different meaning.

Author: No, the intent is definitely that Nix would support previous editions (though maybe not forever). You only get an error if you use a flake that uses an edition more recent than what your version of Nix supports.

Reviewer 32: I think this should be optional for the first edition, since it's just line noise if we end up never breaking compatibility.

  syntax/semantics to be used. This allows the interpretation of
  flakes to change in the future. It also enables some evolution of
  the Nix language; for example, the Nix files in the flake could be
  parsed using a syntax determined by the edition. The only currently

>   allowed value is `201906`. Nix rejects flakes with an unsupported

Reviewer 33: Why use YYYYMM instead of just a number starting at `0` or `1`?

Reviewer 31: It does have the advantage that you can immediately see the age and that parsing it as integer version is strictly monotonic.

Reviewer 13: I'm not sure the age conveys any meaning (vanity metric) and there doesn't seem to be any requirement for versions to be non-monotonic. KISS?

Author: I did think about using a simple counter, but using a date seems nicer to users. For example `edition = 3` doesn't really tell you anything unless you've memorized what each edition is. Whereas I more or less know what's in C++11 or Rust 2019.

It's questionably whether the month is really useful though. If we're sure we're never going to change flakes or the language more than once a year we could drop it :-)

Reviewer 34: We can assume the value of the epoch is a non-monotonic, opaque integer which always increases.

Reviewer 25: Surely it'd be a *monotonic*, strictly increasing, opaque integer?

Reviewer 35: My experience from french as well as my reading of the wiktionary both tell me that “strictly increasing” already implies “monotonic”, and that this word is thus jargon that would be better replaced by just “strictly increasing” for anything related to this, as it's both more precise and easier to understand. (something that always decreases or stays constant is monotonic too)

  edition.

* `name`: An identifier for the flake, used to reference it from
  `inputs`. (E.g. the `nixpkgs` in `inputs.nixpkgs` refers to the

>   identifier of the Nixpkgs flake.)

Reviewer 31: That plays into the flat/nested dependency problem, aka do you need conflict resolution (only one version of a dep allowed at a time) or do you build a tree.

>   identifier of the Nixpkgs flake.)

Author: Currently you can't have more than one flake with the same name. However I've been thinking about changing inputs to an attrset, e.g.
```
inputs = {
  nixpkgs = "github:NixOS/nixpkgs"; # == "nixpkgs"
  nixpkgs_old = "github:NixOS/nixpkgs/nixos-18.03";
};
```

* `description`: A short description of the flake.

* `inputs`: The dependencies of the flake, as a list of flake
  references (described below).

* `outputs`: A function that, given an attribute set containing the
  outputs of each of the input flakes keyed by their identifier,
  yields the Nix values provided by this flake. Thus, in the example
  above, `inputs.nixpkgs` contains the result of the call to the
  `outputs` function of the `nixpkgs` flake, and
  `inputs.nixpkgs.packages.fuse` refers to the `packages.fuse` output
  attribute of `nixpkgs`.

>   In addition to the outputs of each input, each input in `inputs`

Reviewer 36: As far as I understand `flakes` doesn’t support circular dependencies, which I believe is a good thing, but would be good pointing out somewhere

Author: Might be nice to support though :thinking: 

At least we should check for it. I think currently Nix will just go into an infinite recursion...

Author: https://github.com/NixOS/nix/issues/2997

>   also contains some metadata about the inputs. These are:

Author: There is a more verbose way to refer to outputs of an input, namely `inputs.<input-name>.outputs.<output-name>` which doesn't contain the metadata attributes.

  * `outPath`: The path in the Nix store of the flake's source
    tree. This means that you could import Nixpkgs in a more
    legacy-ish way by writing

        with import inputs.nixpkgs { system = "x86_64-linux"; };

    since `nixpkgs` still contains a `/default.nix`. In this case we
    bypass its outputs entirely and only use the flake mechanism to
    get its source tree.

  * `rev`: The commit hash of the flake's Git repository.

  * `revCount`: The number of ancestors of the revision `rev`. This is
    not available for `github:...` repositories (see below), since

>     they're fetched as tarballs rather than as Git repositories.

Author: To generate user-friendly and increasing version strings. For example, Nix's Hydra jobs use this to generate version strings like `nix-2.3pre20190712_aa82f8b`.

  * `lastModified`: The commit time of the revision `rev`, in the
    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
    this is available for both Git and GitHub repositories, so it's

>     useful for generating (hopefully) monotonically increasing version

Reviewer 37: AFAIU, the ways monotonicity can be violated are:
- Force pushes
- Desynchronised clocks on different committers' machines
- Timezone differences.

Naïvely, I'd say that the first two cannot be handled and we'll simply have to rely on best practice, and for the third that the dates should always be in UTC — but I have the impression that date/time stuff is always more complicated than I would have thought, so corrections/improvements on that suggestion would be very welcome.

Author: Yes, hence the "hopefully", it's the best that can be done without access to the `revCount`.. Forced pushes would also break the monotonicity of `revCount` BTW. 

The timestamp is in UTC (via `git log --format=%ct`).

    strings.

  The special input named `self` refers to the outputs and source tree
  of *this* flake.

  A number of outputs have a special meaning, as discussed below.

## Well-known outputs

A number of outputs have a specific meaning to Nix or other tools like
Hydra. Currently, these are:

* `packages`: A set of derivations used as a default by most `nix`

>   commands. For example, `nix run nixpkgs:hello` uses the

Reviewer 2: Given that we have `defaultPackage`, perhaps we should have `defaultOutput` to allow this to be configurable?

Reviewer 31: Alternatively, none of the above, like suggested in https://github.com/NixOS/rfcs/pull/49#discussion_r304090706

Reviewer 2: I'd also be happy with getting rid of both.

  `packages.hello` attribute of the `nixpkgs` flake. It cannot contain

>   any non-derivation attributes. This also means it cannot be a nested

Reviewer 36: Will this be enforced, or will this simply not be recognized by `flake` analysis tools ?

Author: `nix flake check` will complain about them.

  set! (The rationale is that supporting nested sets requires Nix to
  evaluate each attribute in the set, just to discover which packages

>   are provided.)

Reviewer 33: So things like `pythonPackages` won't work anymore?

Reviewer 2: This is awkward for flakes that want to provide many outputs structured in some way, e.g. `haskellPackages`, `docs`, `developmentTools` etc. It sounds like currently those would have to be non-`packages` outputs, which reduces the usefulness of `packages` quite a lot.

Here's an alternative suggestion: allow users to provide an "index filter". Much like other filters, this would be applied at each step of looking for packages to decide whether to index the derivations inside that attribute set. The default would be "packages" (and not any child attribute sets!) but a user could specify a more liberal filter if they wanted (which would let you include attribute sets inside `packages`, or maybe even in other outputs).

Reviewer 4: > supporting nested sets requires Nix to
> evaluate each attribute in the set, just to discover which packages
> are provided.

Why?

Author: When you encounter a attribute like `pythonPackages`, the only way to know that it's an attrset containing more packages rather than a package is by evaluating it.

Reviewer 38: However turning `haskellPackages` into a flake itself would mean we don't need nesting right?

Reviewer 33: There would likely be circular dependencies between `haskellPackages` and the rest of `nixpkgs` though.

Reviewer: Given that flake UI already requires a complicated variable amount of colons, maybe have something like `subflakes` so that `github:NixOS/nixpkgs:pythonPackages:pygtk` could work?

> * `defaultPackage`: A derivation used as a default by most `nix`

Reviewer 33: I think the `defaultPackage`/`defaultApp` can lead to confusion. It can be difficult for newcomers to find out if they have to do
```
nix build dwarffs
nix build dwarffs:dwarffs
nix build nixpkgs:dwarffs
```

I think it would be much clearer if we skip the default packages and just always build `dwarffs:dwarffs` or maybe `dwarffs:pkg`. I think single-usage flakes should be the exception rather than the rule anyway. There is a lot of value in the monorepo approach.

Reviewer 31: Weak +1

Author: I think it's the other way around: most flakes will be single-package (just like most crates are single-app/library). Flakes like Nixpkgs would be the exception

Author: I think it would be bad for UX not to offer the possibility of defaults. Without them, for example, you wouldn't be able to do
```
# nix build
```
anymore. Instead you would have to write
```
# nix build .:<some attribute name that you have to figure out / remember>
```

Reviewer 2: This seems like a feature that we could add later. That way we can see how people actually use it, and whether lots of people want it.

Reviewer 4: I think it’s inevitable that a convention will emerge. If we don’t add it from the start, though, we risk coming up with competing conventions, and it would be nice to avoid that.

Reviewer 37: I already know I want it :)

Author: I know I use it :-)

  commands if no attribute is specified. For example, `nix run
  dwarffs` uses the `defaultPackage` attribute of the `dwarffs` flake.

* `checks`: A non-nested set of derivations built by the `nix flake

>   check` command, or by Hydra if a flake does not have a `hydraJobs`

Reviewer 37: nit: s/or/and/

>   attribute.

Author: An arbitrary derivation that must build successfully in order for `nix flake check` to succeed.

>   attribute.

Author: `hydraJobs` is Hydra-specific. Usually you should use `checks`.

Also, `hydraJobs` can be nested, which means it's impossible to efficiently query what jobs are defined by it. Hence `checks` cannot be nested.

> * `hydraJobs`: A nested set of derivations built by Hydra.

Reviewer 13: What's special about these? Could it be Hydra agnostic like `jobs`? Although I'm not sure that term means much outside Hydra, is it supposed to replace `release.nix`? Maybe `ci`?

Author: The CI-agnostic attribute is `checks`. `hydraJobs` is more of a legacy thing for people who want nested attrsets (like `nixos.tests.installer.x86_64-linux`).

* `devShell`: A derivation that defines the shell environment used by
  `nix dev-shell` if no specific attribute is given. If it does not

>   exist, then `nix dev-shell` will use `defaultPackage`.

Reviewer 37: Is `nix dev-shell` for developing the package itself, like `nix-shell '<nixpkgs>' -A hello`, or for developing _using_ the package, like `nix-shell -p pkgconfig`?

Author: `nix dev-shell` is for development (hence the name), while `nix run` is for using.

So `nix dev-shell` replaces `nix-shell`, while `nix run` replaces `nix-shell -p`.

Reviewer 39: Also `!#/usr/bin/env nix dev-shell` might not work as a shebang. We'll probably still need some top-level shortcut for it.

Reviewer 31: > Also `!#/usr/bin/env nix dev-shell` might not work as a shebang.

It does not work. Never versions of GNU `env` has support for word splitting, but tough luck if you want to be compatible.

Author: https://github.com/NixOS/nix/issues/3046

* `apps`: A set of app definitions used by the `nix app` command. For
  example, `nix app blender-bin:blender_2_79` uses the

>   `apps.blender_2_79` output of the `blender-bin` flake.

Reviewer 37: Are there relevant docs for this? My impression is that this and potentially also `devShell` could be left out of this particular RFC.

Author: Yeah, the `nix app` thing is orthogonal, I just wanted to mention the existence of the outputs (since e.g. `nix flake check` validates them).

Reviewer 4: I think this should be clearer in the RFC, lest other readers be equally confused.

>   `apps.blender_2_79` output of the `blender-bin` flake.

Author: I've removed mention of apps.

* `defaultApp`: An app definition used by the `nix app` command when no
  specific attribute is given. For example, `nix app blender-bin` uses
  the `defaultApp` output of the `blender-bin` flake.

TODO: NixOS-related outputs such as `nixosModules` and `nixosSystems`.

## Flake references

Flake references are a vaguely URL-like syntax to specify which flake
to use. This is used on the command line (e.g. in `nix build
nixpkgs:hello`, `nixpkgs` is a flake reference), and in the list of
flake dependencies in `flake.nix` (e.g. in `inputs = [ "nixpkgs" ];`).

Currently the following types of flake references are supported:

* Git repositories. These have the form

>       (http|https|ssh|git|file):(//<server>)?<path>(\?<params>)?

Reviewer 40: ```suggestion
      (git+http|git+https|git+ssh|git|git+file):(//<server>)?<path>(\?<params>)?
```
It might be useful to use a prefix to avoid confusion in the future on the method of access.

>   with the constraint that `<path>` must end with `.git` for

Reviewer 37: Why this constraint? I frequently use remote non-bare repos via ssh, for instance, and github and gitlab support cloning without a trailing `.git`.

Author: It's to support tarball URLs in the future, e.g. `https://example.org/channel.tar.xz`. This requires disambiguation of Git URLs in some way. An alternative would be to mark them in the URI scheme, like `ssh+git://bla`.

Reviewer 37: I'd prefer marking them in the URI scheme, probably with a prefix rather than a suffix (like pip does, `git+ssh://my-server/nixpkgs`).

Reviewer 4: I agree with @profpatsch (https://github.com/NixOS/rfcs/pull/49#discussion_r303998372) that we should use an attribute we should use an attribute set rather than inventing our own URL format.

Author: https://github.com/NixOS/nix/issues/3045

  non-`file` repositories. `<params>` are a list of key/value pairs in
  URI query parameter syntax. The following parameters are supported:

  * `ref`: The branch or tag to fetch. The default is `master`.
  * `rev`: The Git commit hash to fetch. Note that this commit must be
    an ancestor of `ref`, since Nix doesn't clone the entire
    repository, only the specified `ref` (and the Git protocol doesn't

>     allow fetching a `rev` without a known `ref`).

Reviewer 37: + The default is the commit pointed to by `ref`.

or something to that effect?

  * `dir`: The subdirectory of the repository in which `flake.nix` is
    located. This parameter enables having multiple flakes in a

>     repository. The default is the root directory.

Author: Probably as a matter of policy we don't want to emphasize the existence of this feature since it's bad UX if users have to specify subdirectories.

>     repository. The default is the root directory.

Reviewer 2: I expect it to be pretty common for people to want to have more than one flake in a repository. Given that afaict the flake *has* to be called `flake.nix`, there is no way to do that apart from putting them in subdirectories.

  For example, the following are valid Git flake references:

  * `https://example.org/my/repo.git`
  * `https://example.org/my/repo.git?dir=flake1`
  * `ssh://git@github.com:NixOS/nix.git?ref=v1.2.3`
  * `git://github.com/edolstra/dwarffs.git?ref=unstable&rev=e486d8d40e626a20e06d792db8cc5ac5aba9a5b4`

>   * `file:///home/my-user/some-repo/some-repo.git`

Reviewer 40: I'm not a big fan of URL-stuffing either. The only advantage is that it's a single string that can be copied around.

* Local paths. These have the form

      <path>(\?<params)?

  where `<path>` must refer to (a subdirectory of) a Git repository.

>   These differ from `file://` flake references in a few ways:

Reviewer 37: +1 we already have this problem with nix's `file:///path` stores (flat-file binary caches) vs `/path` (chroot store).

  commit hash without specifying a branch or tag.

  The only supported parameter is `dir` (see above).

  Some examples:

  * `github:edolstra/dwarffs`
  * `github:edolstra/dwarffs/unstable`
  * `github:edolstra/dwarffs/d3f2baba8f425779026c6ec04021b2e927f61e31`

* Indirections through the flake registry. These have the form

      <flake-id>(/<rev-or-ref>(/rev)?)?

  These perform a lookup of `<flake-id>` in the flake registry. The
  specified `rev` and/or `ref` are then merged with the entry in the
  registry. (See below.) For example, `nixpkgs` and
  `nixpkgs/release-19.03` are indirect flake references.

In the future, we should also add tarball flake references
(e.g. `https://example.org/nixpkgs.tar.xz`). It would also be
straight-forward to add flake references to Mercurial repositories
since Nix already has Mercurial support.

## Flake registries

Flake registries map symbolic flake identifiers (e.g. `nixpkgs`) to
"direct" flake references (i.e. any type of flake reference that's not
an indirection). This is a convenience to users, allowing them to do

    nix run nixpkgs:hello

rather than

    nix run github:NixOS/nixpkgs:hello

There are multiple registries:

> * The global registry

Reviewer 4: Why do we need a global registry at all? Could we not just, eg, provide hard-coded defaults for `nix` and `nixpkgs`, and let users do the rest? This would avoid becoming a package repository to any extent, which I think would be a very good thing.

Reviewer 41: @alyssais what would the defaults be?

Reviewer 4: Maybe

- `nix = https://github.com/NixOS/nix stable`
- `nixpkgs = https://github.com/NixOS/nixpkgs unstable`

Author: We don't *need* a registry, but it seems important for UX. In fact it's more important here than in tools like Cargo, since the registry is used in a command line interface. We don't want to require people to type `nix run github:NixOS/patchelf` instead of `nix run patchelf`. Special-casing `nixpkgs` seems ugly, and it makes every other flake second-class compared to `nixpkgs` (which is bad since we should move away from the monorepo approach since it doesn't scale).

Reviewer 33: > which is bad since we should move away from the monorepo approach since it doesn't scale

I disagree here, which is probably also the reason I'm a bit hesitant with the flakes approach in general. For me the monorepo approach is one of the major selling points of nixpkgs. Probably *the* selling point. It lowers the barrier to contribution a lot. It makes it very easy to find any information I want with a single grep. I can bisect any problem. I can test if my update breaks any known reverse dependencies. There is a central point of trust, even if that trust is very diffused at the moment.

I imagine a bunch of single-package repos loosely tied together with lockfiles will actually scale less. Why would I want `patchelf` to be in its own flake vs. in `nixpkgs`?

Reviewer 32: I think a global registry as proposed is a bad idea and should be avoided. That doesn't mean we can't have a local registry, much as channels are registered now. `nix` and `nixpkgs` could be registered by default in the local registry file when Nix is installed.

I'm alarmed by discussion of moving away from a monorepo; I agree that this is a major selling point of NixOS.

Reviewer 42: I would also like to add that "special case `nixpkgs` flake in the UX" is not the only option. I see no reason the user cannot also specify the default flake that resolution should apply to, or alternatively, no default, requiring explicit flake name references. I don't think this would be surprising or unreasonable to support, and would make very obvious use cases easier to support.

For example, when I run `nix run hello`, what if I want the stable branch? Right now from my reading, `nix run nixpkgs/release-19.03:hello` would be used to do this, but if we're talking about UX vs `nix run hello`, I think I think having to do this incantation is generally worse than simply having to write `nixpkgs:hello` for the sake of flake-ambiguity. But again I feel this is a false dichotomy. Of course, I could override the flake registry file with the proposal as it stands, and change the `nixpkgs` flake to use `?ref=release-19.03`, but then I'm basically cloning the entire registry to tweak one setting (so what if it falls out of date?). What if I simply *do* want `nix run hello` to use the stable branch? What if I don't want *any* default? I don't think we should dictate project policy in this manner based on what we _think_, _might be_, a UX concern.

Of course by having user-local flake "repositories", one could argue we are re-introducing a source of in-determinism (no two users necessarily have the same local flake config). But by the other points in the design, flakes are strictly more auditable and easier to manage anyway: they more easily accounted for than `NIX_PATH` search includes or `getEnv` nonsense when debugging builds, etc. A large amount of nonsense can be traced back to stray search includes for example. Asking the user to `cat ~/.config/nix/flakes.json` when debugging something seems far superior than what we have now, global registries or not.

---

Finally, I am also wary of any discussion to move away from a monorepository. "It does not scale" is not evidently true to me in any way. Considering we are one of the most traffic'd repos on all of GitHub and other infrastructural priorities, I'd say it's actually one of our lesser concerns. The monorepo approach of Nix is a huge advantage which I consistently tell people about: it means making atomic changes across components is viable and feasible, that all source code and context (namely, `git` history) is locally available, and that we all have a singular place to coordinate and discuss (mostly). One of the costs we pay, however, and the one that upsets people (from my impressions) in our current model, is that our reliance on tools like `hackage2nix` causes huge amount of "commit object bloat" with +10,000 line diffs and large commit objects. But I am quite willing to argue that the proposed cure, in this case, is *far, far* worse than the disease could ever be.

On top of this, even assuming the Flake RFC goes through -- predicating *any* features or design decisions for Flakes on the *future intention* to "fix" the monorepo is inherently suspect to me. It has not been decided or discussed widely, and I actually suspect the pushback to such a change would be *far* greater than Flakes could ever provoke. It is not a factor that should be considered in the discussions or design of the Flake tooling or UX, as far as I'm concerned. There are plenty of other benefits to think about for now.

Reviewer 13: One use case to consider are companies, where it could be monorepo or polyrepo.

Currently registry input is non-deterministic, so one user could have different version than the other, resulting in different lock output.

For the polyrepo case I suggest:

- registry becomes a nix file, so that it can be overriden/expanded
- adding registry field to flakes like `registry = globalRegistry: (self: super: { extraIdentifier = ...; })` with a default value of `id` function

That's an alternative way to achieve https://github.com/NixOS/rfcs/pull/49/files#r304091661

Reviewer: Whatever we do with `haskellPackages`, the problem we want to solve is «checking in huge autogenerated files»; if we solve this, the main scaling problem of monorepo — deciding when to pin old dependency versions — is at least more transparent with a monorepo, and needs to be solved somehow in any layout.

Splitting the repo _and_ collecting it back via flake registry sounds like «NixOS/Nixpkgs repository split», only larger scale; we actually moved towards monorepo after the move from SVN broke explicit synchronisation of the two subrepositories.

Reviewer 4: Thought: these problems would potentially be solved if Nixpkgs were a monorepo of flakes. Stuff like Haskell generated files could be fetched from elsewhere as a flake, but locked to a specific version.

I haven’t thought in depth yet about how this would work with lockfiles, etc.

Reviewer 25: @alyssais, isn't that exactly what @7c6f434c mentions in their second paragraph as being problematic?

Reviewer: @toonn Not exactly — the top-level repository with subtrees does provide a sequence of «cuts» with exact commit choice for all parts; this, for example, provides easier checkouts of historical states (and also that elusive state «what people are considering master branch head right now») than fully independent repos. Separate NixOS and Nixpkgs repos hoted on GitHub lacked full tooling support for this.

Author: Regarding monorepos vs polyrepos: flakes make a polyrepo world easier but don't mandate it. So a decision to move away from a monorepo Nixpkgs/NixOS is separate from this RFC. Ultimately however the monorepo approach fundamentally doesn't scale: it shouldn't be necessary for everybody who just wants to share a NixOS module to have to add it to the `nixpkgs` repo. Note by the way that any NixOS system is already a polyrepo configuration because a system's `configuration.nix` is not part of the nixpkgs repo. But Nix at the moment lacks a good way to manage that dependency.

Regarding registries: the UX value of global registries is obvious, IMHO, from tools like Cargo or npm.

Slightly tangentially, here is a post [1] discussing `crates.io` policies, which would be useful for the flake registry as well. It also discusses why `crates.io` doesn't have namespaced packages.

[1] https://internals.rust-lang.org/t/crates-io-package-policies/1041/1

Author: > Why would I want patchelf to be in its own flake vs. in nixpkgs?

That's actually a good example :-) The reason you'd want that is because the patchelf package in `nixpkgs` is several years behind the upstream package. With patchelf-as-a-flake you're no longer dependent on a single monorepo.

Reviewer 4: > Regarding registries: the UX value of global registries is obvious, IMHO, from tools like Cargo or npm.

npm’s global registry has been a disaster. There’s been account compromises, typosquatting, name squatting in general, and useful names lie dormant because somebody grabbed them years ago.

I am strongly opposed to a global registry, especially to start with. If somebody wants to start a global registry outside of Nix to show its value, fair enough. Maybe we can adopt it. But we don’t need it for flakes, so I don’t think we should do it now.

Reviewer 4: A search engine for flakes might be a nice idea, however. Again, though, it can be external to Nix.

Reviewer 33: > The reason you'd want that is because the patchelf package in nixpkgs is several years behind the upstream package

In this case why don't we update the nixpkgs version instead?

Author: crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

@timokau Patchelf was just an example, but in general you can't expect Nixpkgs to contain the whole world.

Reviewer 43: Name squatting is also an issue on crates.io, unfortunately. Now, it doesn't have to be a disaster, but it certainly needs moderation and, ideally manual, verification before flake submissions. If the official global registry is small, and users can just add third-party registries to their systems, it shouldn't be bad. Namespacing would also help, but I feel like the official global registry should be limited to official NixOS stuff anyways.

Reviewer 4: > crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

You could have the same ergonomics with a local registry only, with the nice commands for adding to it proposed in this rfc, though, with none of the need for us to maintain any sort of global registry.

Reviewer 4: > Ultimately however the monorepo approach fundamentally doesn't scale: it shouldn't be necessary for everybody who just wants to share a NixOS module to have to add it to the nixpkgs repo.

I don’t want to argue this further in this thread, because I think it’s off-topic for the RFC, but I want to make clear that I still don’t agree and any such move would need to be discussed extensively at a later date before it were to happen.

Reviewer 44: > crates.io works very well so I don't think a registry needs to be a disaster. In any case,you don't have to use the registry. Of course then you have to type things like `nix build github:NixOS/nixpkgs:...` instead of `nixpkgs:...` all the time.

You mentioned crates.io a lot in your NixCon talk and it's obvious that this RFC took a lot of inspiration of what Rust is doing. I'd like to chime in here that the public often doesn't see a lot of the effort and sweat that it takes on parts of *people* to run crates.io, and the architectural problems that it comes with. This doesn't just include the actual infrastructure and the team of people it takes to run it, but also a lot of policy decisions on things like DMCA requests, name squatting, etc that introduce a lot of work that general users of the service aren't aware of either. It's easy to say "look, it runs really well" and not see the weekly meetings on policy updates. I got the feeling that everybody involved in developing NixOS was already pretty booked out. I feel that adding yet another thing to maintain to the list of NixOS infrastructure things would be an extremely bad idea.

Reviewer 44: The `crates.io` ecosystem is **fundamentally** a different maintenance approach to NixOS. There are no "Rust maintainers". Sure, there's people who work on the compiler or who are active in various working groups and teams (:wave:) but there's no centralised development in the way that NixOS requires it. Building a bunch of libraries is different from building an operating system distribution and while a global registry of libraries works well for one, it wouldn't scale at all for the other because it's the same people involved in most of the pieces. It just adds overhead.

Reviewer 44: **Re: the mono repo**

As others have said previously, the fact that all the moving parts in NixOS are in a single place, can easily be bisected and tested against one another is a **huge** selling point for collaboration and also lowers the barrier to entry. NixOS is the first Linux distribution I've actually started working on, despite wanting and very much trying to get involved in both Debian and Fedora. I think it would be a shame if this were to change.

There's a relevant discussion/story from the Rust/ `rustc` repo about this. The repo itself is huge, unwieldy, takes ages to download etc. So it was proposed to remove certain components from it that are only tangentially related to the compiler core, such as `rustdoc`. Ultimately, it was decided to keep the repo as it is, because `rustdoc` is a component that frequently breaks with new PRs and so, it's common practise that if a compiler contribution breaks docs, that person then also fixes docs.

I think you could compare the development of `rustc` with the development of `NixOS`/`nixpkgs`. They are both mono repos, both part of an ecosystem with moving pieces around it, but strongly integrated i each other and benefit greatly from having all their parts kept in the same place.

  https://raw.githubusercontent.com/NixOS/flake-registry/master/flake-registry.json. (This
  location can be overriden via the `flake-registry` option.) Nix
  automatically fetches this registry periodically. The check interval

>   is determined by the `tarball-ttl` option.

Reviewer 37: While it makes sense in terms of the resulting behaviour, I'm not a fan of reusing options for purposes that don't match their names. I'd prefer introducing a new option (`registry-cache-ttl`?) for this, or renaming the `tarball-ttl` option.

Author: :+1: on renaming.

Author: https://github.com/NixOS/nix/issues/3044

> * The local registry `~/.config/nix/registry.json`.

Reviewer 4: Should be `$XDG_CONFIG_HOME/nix/registry.json`.

Reviewer 37: Or local _registries_, searching `$XDG_CONFIG_PATH` for `nix/registry.json`. That allows different levels of the config, e.g. system-wide with user overrides.

Reviewer 31: Though it shouldn’t be so complicated that it’s easy to miss configuration. So keeping it simple is key for less surprises.

Reviewer: Maybe a single directory with multiple registries and a debug command to list the local registries?

I also think that allowing only a single local-FS registry can create annoyance in too many ways… Neither `$NIX_PATH` not overlays are that restrictive.

Author: The implementation uses `$XDG_CONFIG_HOME` in fact, but I didn't want to mention that here for simplicity. 

@lheckemann I guess you meant `$XDG_CONFIG_DIRS`? That would be pretty useful.

Reviewer 37: I did indeed :) and it's good for consistency, since we have that behaviour for nix.conf already with nixos/nix#2501.

> * The local registry `~/.config/nix/registry.json`.

Reviewer 13: I'd appreciate if there would be support for `registry.json` local to `flake.nix`. 

That way git repository can ship the registry instead of specifying how to create one or require to be in global registry (not always feasible as mentioned elsewhere).

* Registry entries specified on the command line.

> A registry is a JSON file that looks like this:

Reviewer 4: Why not Nix? Using JSON and a management CLI rather than encouraging editing the file for this feels a bit at odds with how Nix does everything else.

Reviewer 31: Nix is turing complete, disallowing (recursive) functions has a lot of merits to disallow whole classes of problems.

Reviewer 40: It's more that Nix code is hard to patch with new values if there are functions in there. If it was a subset of Nix with only the Object Notation then it would be easy.

Reviewer 36: However that still means we’d need nix-specific tools to do any kind of automated housekeeping, while there’s already plenty of those for JSON

Reviewer 4: If we want automated housekeeping, I’m all for JSON. The question I’m asking is “do we?”.

Author: To turn it around: what's the argument for making it a Nix file?

Reviewer 45: > Nix is turing complete, disallowing (recursive) functions has a lot of merits to disallow whole classes of problems.
True, but this affects a large part of the nix ecosystem beyond just a registry.
>Nix code is hard to patch with new values
I don't believe there exists anywhere else where Nix will overwrite a nix file that you have written. Nix shouldn't be changing your Nix files.
>what's the argument for making it a Nix file
The configuration for the entire rest of the system, bar .nix-channels, is configured using Nix. It's just weird that the channels, or this flake system is using an entirely separate configuration from the rest of the system. This makes those systems not feel first class as you have to configure them separately. Personally, I would rather see these systems getting added to the actual configuration files for the system (/etc/nixos/configuration.nix) or users (like home-manager's home.nix). Having the configuration split up into it's own little system as a .nix-channels or a registry.json just complicates reproducing machines. It feels like it's a mutable part to the system being able to change your channels which can cause the same nix file create two different systems based off what channel and what version of it you are on. I will admit the drawback to doing it this way is that it is harder to upgrade as you'll need to update the branches and which revision of them you want your system to use. Perhaps it may be worth making a trade off between reproducability and convenience by allowing the version of the channel you want your system to be on to be optional (yes I know this is similar to the idea of pinning).

Reviewer 13: The upside of using Nix is that registries can be override/expanded. The downside is that it's harder to generate them. Should we generate registries? I think we'd only need their output, which can be JSON.

Author: I think for the flake MVP we should stick to JSON. We can always introduce `.nix` registries in the future if there is demand, but since they're simple `identifier -> flake location` mappings, I don't think it will be needed.

```
{
    "version": 1,
    "flakes": {
        "nixpkgs": {
            "uri": "github:NixOS/nixpkgs"
        },
        ...
    }
}
```

With this registry, flake references resolve as follows:

* `nixpkgs` -> `github:NixOS/nixpkgs`
* `nixpkgs/release-19.03` -> `github:NixOS/nixpkgs/release-19.03`
* `nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c` -> `github:NixOS/nixpkgs/f1c995e694685d6dfb877f6428d3e050d30e253c`

The registries are searched in reverse order. Thus the local registry
overrides the global registry, and the command line takes precedence

> over the local and global registries.

Reviewer 40: probably the same as:
```
nix-repl> builtins.fromJSON ''{ "foo": 3, "foo": 4 }''
{ foo = 4; }
```

> over the local and global registries.

Reviewer 31: Should definitely be documented!

> over the local and global registries.

Reviewer 46: Then `"flakes"` should be an array of objects in JSON terms, because objects are unorderdered and some implementations like databases will reorder the entries.

as experimental, so we have some freedom to make incompatible
changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
`nix-instantiate`) should not be changed to avoid breakage.

Most `nix` subcommands work on a list of arguments called
"installables" for lack of a better word. For example,

    # nix run nixpkgs:hello dwarffs:dwarffs

takes two flake-based installables. The general form is:

    <flake-ref>(:<attr-path>)?

Examples of installables:

* `nixpkgs:packages.hello`
* `nixpkgs:hello` - short for `nixpkgs:packages.hello`
* `nixpkgs/release-19.03:hello` - overrides the Git branch to use
* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082:hello` -
  specifies the exact Git revision to use
* `dwarffs` - short for `dwarffs:defaultPackage`
* `nix:hydraJobs.build.x86_64-darwin`
* `.:hydraJobs.build.x86_64-darwin` - refers to the flake in the
  current directory (which can be a dirty Git tree)
* `.` - short for `.:defaultPackage`

If no argument is given, the default is `.`; thus,

    # nix build

is equivalent to

    # nix build .:defaultPackage

For backwards compatibility, it's possible to use non-flake Nix

> expressions using `-f`, e.g. `nix build -f foo.nix foo.bar`.

Reviewer 37: -1 on putting backwards compatible cruft in a tool that we still have the freedom to change! AFAIK, the only thing the current `nix build` can do which `nix-build` can't is display a fancy progress bar, so IMHO `nix-build` should be preferred for non-flake expressions.

Reviewer 47: Is there a consensus on deprecating or continuing the support of the "nix-*" commands? I prefer using the `nix` command because its arguments are consistent and the UX matches other tools such as git and cargo.

If we take your suggestion, should the `-f` argument be removed from other `nix` commands, such as `nix eval`? I personally think doing so would reduce `nix`'s usability as a general-purpose build tool rather than as a flake builder.

Reviewer 31: > Is there a consensus on deprecating or continuing the support of the "nix-*" commands? I prefer using the `nix` command because its arguments are consistent and the UX matches other tools such as git and cargo.

The `nix` command is not even stable yet, so nobody should use them for scripting. We can think about starting deprecation once we give any stability guarantees.

Author: I don't think we should remove `-f` because it's unrealistic to think that everything will be converted to flakes right away, and there are non-legacy uses of `-f` (e.g. `nix eval -f ...`).

## Lock files and reproducible evaluation

Inputs specified in `flake.nix` are typically "unlocked": they are
flake references that don't specify an exact revision (e.g. `nixpkgs`
rather than
`github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082`). To
ensure reproducibility, Nix will automatically generate and use a
*lock file* called `flake.lock` in the flake's directory. The lock
file contains a tree of mappings from the flake references specified
in `flake.nix` to direct flake references that contain revisions.

For example, if `flake.nix` contains
```
inputs =
  [ "nixpkgs"
    github:edolstra/import-cargo
  ];
```

> then the resulting lock file might be:

Reviewer 4: I’ve always thought of derivations as a kind of lockfile — I suppose that wouldn’t make sense here?

Author: You mean .drv files? Yeah I suppose they're a kind of lockfile, but we can't really use them here because they're too general (encoding arbitrary builds) and would require a lot of small files.

```
{
    "version": 2,
    "inputs": {
        "github:edolstra/import-cargo": {
            "id": "import-cargo",
            "inputs": {},
            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Author: It's because the `github:` scheme has some particular caching semantics, in particular using the fact that the `ETag` returned by GitHub is the Git revision. We can't assume that for `https` tarballs in general.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 4: We should come up with a better name for this scheme, then, and make it so the full URL is still included. I don’t think Nix should be coupled to GitHub (and as proposed this wouldn’t even work with GitHub enterprise). If we used attribute sets instead of URLs, this could be an attribute in the set.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Author: Hm, why wouldn't it work with GitHub enterprise?

There is definitely something to be said for attrsets instead of URLs, but it means that we would have different syntaxes for flake references on the command line and in `.inputs`.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 4: Well, there’s no way to specify the domain GitHub is running on.

The command line is a good point. Could just be multiple arguments, maybe?

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 32: IMO, the best option here is to abandon the `github:` scheme and specially recognise URLs of the form `https://github.com/...` as supporting an alternate retrieval strategy. The advantages of this would be:

- Separation of *what* is retrieved from *how* it is retrieved; `github:` and `https://github.com/...` are the same *what*, just being retrieved via a different method. But an URL is supposed to specify the what, not the how.
- This avoids having to change `github:` URLs if the retrieval methods GitHub supports change.
- Conversely, suppose some other hosting site `https://example.com/` becomes popular in the future. If you want to add a more efficient way of retrieving Git repositories, with this model you'd be asking them to change to `example:...` to do so, which would break compatibility with older versions of Nix which don't understand `example:...`.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 25: Shouldn't it be up to the user whether they want to use https or something else? Similar to how you get the option to clone using https or ssh in the github interface.

While a URI does indeed merely identify a resource a URL definitely most commonly specifies a protocol to be used for the retrieval. So I'd rather see a scheme where `git://github.com/...` is valid than having https urls being subverted non-obviously.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Author: > IMO, the best option here is to abandon the github: scheme and specially recognise URLs of the form https://github.com/... as supporting an alternate retrieval strategy.

Special-casing certain HTTP URLs seems *much* uglier than having a separate URI scheme. You would expect all resources in a particular URI scheme to be interpreted uniformly.

Also, from a UX perspective, having to figure out / type `https://github.com/edolstra/patchelf/archive/master.tar.gz` (and hoping that that's the special URL recognized by Nix) is worse than `github:edolstra/patchelf`.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 4: Special casing would be a very bad idea, because GitHub URLs don’t support all the features of git URLs. Like many other things in this rfc, I think we should be asking ourselves “is this feature absolutely critical for an initial release?” The answer to me in this case is quite clearly no. So I propose we drop GitHub URLs for now, in favour of git URLs, pending further, more focussed discussion.

>             "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"

Reviewer 48: Really late, but couldn't this quite easily be solved with something like github:(customdomain.com)edolstra/patchelf?

        },
        "nixpkgs": {
            "id": "nixpkgs",
            "inputs": {},
            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
        }
    }
}
```

Thus, when we build this flake, the input `nixpkgs` is mapped to
`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix

> will also check that the content hash of the input is equal to the one

Reviewer 36: I’m guessing that’s the `narHash`? Wasn’t entirely clear from the paragraph above

Author: No, it's the Git revision.

recorded in the lock file. This check is superfluous for Git
repositories (since the commit hash serves a similar purpose), but for
GitHub archives, we cannot directly check that the contents match the
commit hash.

> Note that lock files are only used at top-level: the `flake.lock` files

Author: Yes, good point. We could simply always use the lock file of a dependency if it exist (so good practice would be not to commit lock files for library flakes). It gets tricky for flakes that provide both libraries and applications (like Nixpkgs)...

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 49: This seems like something to consider carefully or it could become a big pain.

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 49: I'm not sure I understand the difference between `packages` and `apps`, but what about having whether a lock file is considered be based on which set the reference is in?

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 12: I think also that when a lock file is present, it should be used. It basically says "we provide a curated set, don't break it". It's then up to the community to ensure they don't use it incorrectly. 

Now, in practice I think that is not something we can always rely on and so we want to be able to override/patch. Also, because upstreams tend to pin things to much; just look at the amount of patching we need with Python applications. We are given the attribute`inputs`, which is already imported, and won't allow us to override or patch the Flake when needed which, if correct, is a limitation.

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 12: @ryantm see e.g. the explanation given for Cargo, https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 49: Oh, right, this would be based on the inputs and `packages`/`apps` are outputs.

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 32: Lockfiles shouldn't be committed to repositories under any circumstance, unless it is a distribution repository. This is no different to [vendor bundling](https://fedoraproject.org/wiki/Bundled_Libraries), which is rightly opposed. It's for a distributor to decide how to combine dependencies.

I agree that lockfiles in dependencies should be ignored. If this principle were abandoned, it should be strictly opt-in.

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 50: Could someone explain why lock files in dependencies should be ignored? Doesn't it break the reproducibility of evaluation if the inputs of the dependency got updated?

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 20: Without care, there is a tension between flakes offering composability on the one hand and reproducibility on the other. Ignoring dependent lockfiles is one way of reducing that tension: The flake.nix should specify the *interface* for the flake (eventually with version bounds or similar?) and the flake.lock should specify the *current composition* for people working on that project directly.

For example, I'd like us to eventually have a `stdenv` flake that abstracts out the bootstrapping logic and definition of mkDerivation from nixpkgs proper. When I'm developing on `stdenv` itself, I want some *fixed* definitions of things like gcc etc. so I have something to work with, but the entire point is that I ultimately want the *caller* to specify these! So when nixpkgs imports the stdenv flake, it should provide the pieces that make it all work.

> Note that lock files are only used at top-level: the `flake.lock` files

Reviewer 35: Maybe it might make sense, for a first draft, to say “lock files in dependencies are a hard error for now” and to just add that as a to-be-resolved-later question?

I'm not sure we need what @FRidh wants right now (and it's something that apparently only makes sense for applications, so not that well-explored for build systems at least), and this way is forwards-compatible with any future road we might end up choosing for handling lock files in dependencies.

in dependencies (if they exist) are ignored.

> Nix automatically creates a `flake.lock` file when you build a local

Reviewer 31: Rust’s `cargo` automatically creates a lock file, and it leads to a nice user experience, because locking the version you are working with is always what you want.

> Nix automatically creates a `flake.lock` file when you build a local

Reviewer 40: My experience is that if the lockfile is not generated by default, developers forget to commit the lockfile. Which then introduces all sort of reproducibility problems. When the lockfile is generated automatically it becomes easy to compare changes on updates.

> Nix automatically creates a `flake.lock` file when you build a local

Reviewer 32: Lockfiles should not be commited anyway as this is effectively no different from [vendor bundling](https://fedoraproject.org/wiki/Bundled_Libraries).

I agree that generation of lockfiles should be opt-in.

> Nix automatically creates a `flake.lock` file when you build a local

Reviewer 40: Unlike bundling, the lockfiles don't prevent people from using a different dependency set. What it does it provide a well know working set that can be used to build the project reproducibly.

> repository (e.g. `nix build /path/to/repo`). It will also update the

Reviewer 37: \+ which does not already have a lock file

?

lock file if inputs are added or removed. You can pass
`--recreate-lock-file` to force Nix to recreate the lock file from
scratch (and thus check for the latest version of each input).

Lock files are not sufficient by themselves to ensure reproducible
evaluation. It is also necessary to prevent certain impurities. In
particular, the `nix` command now defaults to evaluating in "pure"
mode, which means that the following are disallowed:

* Access to files outside of the top-level flake or its inputs, as
  well as paths fetched using `fetchTarball`, `fetchGit` and so on

>   with a commit hash or content hash. In particular this means that

Reviewer 36: Same question, except that I can imagine it be "with only a commit hash" which would mean relying on git to do a good job at hashing contents

  Nixpkgs will not be able to use `~/.config/nixpkgs` anymore.

* Access to the environment. This means that `builtins.getEnv "<var>"`

>   always returns an empty string.

Author: Maybe, but then you would need a way to test whether you're running in pure mode.

* Access to the system type (`builtins.currentSystem`).

* Access to the current time (`builtins.currentTime`).

* Use of the Nix search path (`<...>`); composition must be done
  through flake inputs or `fetchX` builtins.

Pure mode can be disabled by passing `--impure` on the command line.

## Evaluation caching

> The fact that evaluation is now hermetic allows `nix` to cache flake

Reviewer 37: \+ when evaluating purely

At risk of stating the obvious, I'd rather have this in there than end up with the cache being poisoned when I evaluate impurely once :)

attributes. For example (doing `nix build` on an already present
package):

    $ time nix build nixpkgs:firefox
    real    0m1.497s

    $ time nix build nixpkgs:firefox
    real    0m0.052s

The evaluation cache is kept in `~/.cache/nix/eval-cache-v1.sqlite`,
which has entries like

    INSERT INTO Attributes VALUES(
      X'92a907d4efe933af2a46959b082cdff176aa5bfeb47a98fabd234809a67ab195',
      'packages.firefox',
      1,
      '/nix/store/pbalzf8x19hckr8cwdv62rd6g0lqgc38-firefox-67.0.drv /nix/store/g6q0gx0v6xvdnizp8lrcw7c4gdkzana0-firefox-67.0 out');

where the hash `92a9...` is a fingerprint over the flake store path
and the contents of its lockfile. Because flakes are evaluated in pure

> mode, this uniquely identifies the evaluation result.

Reviewer 36: What happens in `—impure` mode? Are those discarded or (similar to (un)sandboxed build) could be kept around nonetheless?

Author: In impure mode caching is disabled.

> mode, this uniquely identifies the evaluation result.

Author: Hm, what do you mean with "resolving"? The existence or non-existence of the cache should not matter in general.

> Currently caching is only done for top-level attributes (e.g. for

Reviewer 37: "top-level attributes" is confusing here — should this be "attributes specified in a command"? Because as far as I can tell, `packages.firefox` is _not_ a top-level attribute.

  For example,

      # nix flake add nixpkgs github:my-repo/my-nixpkgs

  redirects `nixpkgs` to a different repository. Similarly,

      # nix flake add nixpkgs github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757

  pins `nixpkgs` to a specific revision.

  Note that registries only have an effect on flake references used on
  the command line or when lock files are generated or updated.

* `nix flake remove <flake-ref>`: Remove an entry from the local
  registry.

* `nix flake pin <flake-ref>`: Compute a locked flake reference and
  add it to the local registry. For example:

      # nix flake pin nixpkgs

  will add a mapping like `nixpkgs` ->
  `github:NixOS/nixpkgs/444f22ca892a873f76acd88d5d55bdc24ed08757` to
  the local registry.

* `nix flake init`: Create a skeleton `flake.nix` in the current
  directory.

* `nix flake update`: Recreate the lock file from scratch.

* `nix flake check`: Do some checks on the flake (e.g. check that all
  `packages` are really packages), then build the `checks`
  derivations. For example, `nix flake check patchelf` fetches the
  `patchelf` flake, evaluates it and builds it.

> * `nix flake clone`: Do a `git clone` to obtain the source of the

Reviewer 4: Should have a less Git-specific name.

  specified flake, e.g. `nix flake clone dwarffs` will yield a clone

>   of the `dwarffs` repository.

Reviewer 36: And alternatively, once the clone is performed, does the lockfile point to the clone?

>   of the `dwarffs` repository.

Author: The lockfile never points to its own repository (that would be impossible because changing the lockfile invalidates the repo hash). I.e. the lockfile only locks the dependencies, not the top-level flake.

If the flake is in a subdirectory, it will still clone the entire repository. This is necessary because the flake can refer to files in parent directories in the repository.

## Offline use

Since flakes replace channels, we have to make sure that they support
offline use at least as well. Channels only require network access
when you do `nix-channel --update`. By contrast, the `nix` command
will periodically fetch the latest version of the global registry and
of the top-level flake (e.g. `nix build nixpkgs:hello` may cause it to
fetch a new version of Nixpkgs). To make sure things work offline:

* It's not a failure if we can't fetch the registry.

* It's not a failure if we can't fetch the latest version of a
  flake. (For example, `fetchGit` no longer fails if we already have
  the specified repository and `ref`.)

* Fetched flakes are registered as garbage collector roots, so running

>   the garbage collector while on an airplane will not ruin your day.

Reviewer 37: Do we want to specify a mechanism for removing/expiring these roots?

* There is an option `--no-net` to explicitly prevent updating the
  registry or doing other things that need the network (such as
  substituting). This is the default if there are no configured
  non-loopback network interfaces.

## CI integration

Flakes are useful for CI systems since `flake.nix` and the lock file
exactly specify the input repositories. For Hydra this means that you
no longer need to configure any jobset inputs; just a top-level flake
reference is enough. For example, the [configuration of the `patchelf`
jobset](https://hydra.nixos.org/jobset/patchelf/master#tabs-configuration)
just points to `github:NixOS/patchelf` and doesn't need to specify

> `nixpkgs`.

Reviewer 40: Will Hydra be able to push the sources fetched by flake into the binary cache? If I understand it correctly those are fetched at eval-time and only linked into the user's .cache/nix folder.

Author: In principle it can do that. Currently they'll only end up in the binary cache if they are a dependency of a derivation.

# Drawbacks
[drawbacks]: #drawbacks

Pure evaluation breaks certain workflows. In particular, it breaks the
use of the Nixpkgs configuration file. Similarly, there are people who
rely on `$NIX_PATH` to pass configuration data to NixOps
configurations.

> 

Reviewer 37: I don't think this RFC deprecates use of nix for that purpose?

> 

Author: Yeah, there is nothing precluding flake outputs to generate things like Terraform configurations.

# Alternatives
[alternatives]: #alternatives

For composition of multi-repository projects, the main alternative is
to continue on with explicit `fetchGit` / `fetchTarball` calls to pull
in other repositories. However, since there is no explicit listing of
dependencies, this does not provide automatic updating (i.e. there is

> no equivalent of `nix flake update`).

Reviewer: And we even have experience with various autoupdaters in Nixpkgs! If the autoupdate does nothing interesting as it is limited to git repositories, none of the complicated cases would arise.

Instead of a `flake.nix`, flakes could store their metadata in a
simpler format such as JSON or TOML. This avoids the Turing tarpit
where getting flake metadata requires the execution of an arbitrarily
complex, possibly non-terminating program.

# Unresolved questions
[unresolved]: #unresolved-questions

> * How to handle the system type? Currently `x86_64-linux` is

Reviewer 40: Injecting the `localSystem` is probably a necessity to make this practical. Maybe it can be a default argument instead of being attached to a builtin. Then if the user wants to cross-build or switch the localSystem they can always pass `-A localSystem something-else` or `-A crossSystem ...`

> * How to handle the system type? Currently `x86_64-linux` is

Author: One possibility is to extend the flake `outputs` function with a set of arguments, e.g.
```
outputs = { inputs, args }: 
  with import inputs.nixpkgs { system = args.system; };
  { ...
  };
```
The `args.systems` argument would be provided automatically by the user interface. It would alos be part of the cache key so evaluations of the same flake for different systems wouldn't collide with each other.

The alternative is to lazily return packages for all supported platform, e.g.
```
outputs = inputs: {
  packages.hello.x86_64-linux = ...;
  packages.hello.x86_64-darwin = ...;
};
```
(This is similar to what Nixpkg's Hydra jobset does.) Then the UI could automatically select a package for the local system (e.g. `nix install flake:hello` would select `packages.hello.<my-system>`).

> * How to handle the system type? Currently `x86_64-linux` is

Reviewer 23: One piece of positive feedback: I can use uses for both approaches but happily the `with import inputs.nixpkgs { system = ... }` approach works today on Darwin just by hardcoding `system = "x86_64-darwin";`, at least for the simple flake I'm testing with!

  hard-coded everywhere.

* How to do Nixpkgs overlays? In principle, overlays would just be a
  kind of flake output.

* More in general, how to handle flake arguments? This must be done in
  a way that maintains hermetic evaluation and evaluation caching.

* What are the criteria for inclusion in the global flake registry?

* Hammer out the details of NixOS/NixOps support for flakes.

* Currently, if flake dependencies (repositories or branches) get
  deleted, rebuilding the flake may fail. (This is similar to
  `fetchurl` referencing a stale URL.) We need a command to gather all
  flake dependencies and copy them somewhere else (possibly vendor
  them into the repository of the calling flake).

# Future work
[future]: #future-work

* The "edition" feature enables future Nix changes, including language
  changes. For example, changing the parsing of multiline strings
  (https://github.com/NixOS/nix/pull/2490) could be conditional on the

>   flake's edition.

Reviewer 31: This is a good practice. Rust does something similar with [editions](https://doc.rust-lang.org/nightly/edition-guide/editions/index.html).

The key here is that the compiler (or evaluator in nix’s case) never loses the ability to work with older editions.

Reviewer 4: How would parsing these work? With Cargo, it’s not like the TOML format is changing between editions. Does the evaluator have to try evaluating with each version of the syntax, starting with the newest, until it finds one that works?

Reviewer 31: Good question, but do we really want to change the nix syntax in a backwards-incompatible way at this point? I think probably no.

If we don’t intend to change the syntax it’s just about the semantics of the fields, which is what `editions` are about I presume.

Reviewer 33: > Good question, but do we really want to change the nix syntax in a backwards-incompatible way at this point? I think probably no.

There's https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d

Author: @alyssais That's a good point, and is an argument in favor of using JSON/TOML for `flake.nix`.

Reviewer 13: Shouldn't `edition` just be evaluated lazily? That seems like a sane requirement.

Author: It can't be evaluated lazily. It has to be evaluated eagerly because the interpretation of all other attributes (potentially) depends on it.

Reviewer: I guess technically there is also the option of including edition into the main flake file name, like some codebases used `.py2` and `.py3` extensions at some point.

  not nearly as reproducible as Nix builds. Nix expressions can access
  arbitrary files (such as `~/.config/nixpkgs/config.nix`),
  environment variables, and Git repositories. This means for instance
  that it is hard to ensure reproducible evaluation of NixOS or NixOps
  configurations.

* Nix projects lack discoverability and a standard structure. For
  example, it's just convention that a repository has a `release.nix`
  for Hydra jobs and a `default.nix` for packages.

* There is no standard way to compose Nix projects. Typical ways are
  to rely on the Nix search path (e.g. `import <nixpkgs>`) or to use
  `fetchGit` or `fetchTarball`. The former has poor reproducibility,
  while the latter is bad UX because of the need to manually update
  Git hashes to update dependencies.

* `nix-channel` needs a replacement: channels are hard to create,
  users cannot easily pin specific versions of channels, channels
  interact in *ad hoc* ways with the Nix search path, and so on.

The flakes mechanism seeks to address all these problems.

# Detailed design
[design]: #detailed-design

## Flakes

A flake is a Git repository that contains a file named `flake.nix` in

> the root directory. (In the future, there might be other types of

Reviewer 8: Supporting tarballs seems like it's pretty necessary to me, especially since there's no pre-existing requirement to put nix files in a git repository (instead of e.g. mercurial, etc etc). On that point, it seems odd to merge in a possible RFC without even writing down the possibility of something non-git-based; especially since (to me) [it feels like](https://github.com/NixOS/rfcs/pull/49/files#r314726559) the only reason these are not in the spec is because it isn't yet implemented in nix.

depends on the Nixpkgs flake and provides a package (i.e. an
installable derivation) and a NixOS module.

```
{
  edition = 201909;

  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

  outputs = { self, nixpkgs }: rec {
    packages.dwarffs =
      with nixpkgs.packages;
      with nixpkgs.builders;
      with nixpkgs.lib;

      stdenv.mkDerivation {
        name = "dwarffs-0.1.${substring 0 8 self.lastModified}";

        buildInputs = [ fuse nix nlohmann_json boost ];

        NIX_CFLAGS_COMPILE = "-I ${nix.dev}/include/nix -include ${nix.dev}/include/nix/config.h -D_FILE_OFFSET_BITS=64";

        src = self;

        installPhase =
          ''
            mkdir -p $out/bin $out/lib/systemd/system

            cp dwarffs $out/bin/
            ln -s dwarffs $out/bin/mount.fuse.dwarffs

            cp ${./run-dwarffs.mount} $out/lib/systemd/system/run-dwarffs.mount
            cp ${./run-dwarffs.automount} $out/lib/systemd/system/run-dwarffs.automount
          '';
      };

    nixosModules.dwarffs = ...;

    defaultPackage = packages.dwarffs;

    checks.build = packages.dwarffs;
  };
}
```

A flake has the following attributes:

> * `edition`: A number that specifies the version of the flake

Reviewer 44: I know Rust has had a lot of popularity with adding an `edition` flag, and we kinda did that because we wanted to fix things that were just bad in the initial design (namely, the module system which was confusing, and the way that macros are exported, which was more of a technical limitation of the time when `1.0` was stabilised).

I feel that declaring up front that `nix` will have editions, without being very clear about what kind of scope the syntax or semantics changes can include opens you up to "edition creep" of future versions, also potentially over-bloating the nix daemon because it will have to support all of these new editions.

Don't get me wrong: editions can be a great tool to fix mistakes of the past in a way that doesn't break everything that depends on them. But it's a commitment and also a maintenance burden. And without a clear roadmap of **what** should be fixed and **why**, just willy nilly adding an edition flag is, in my opinion, very dangerous!

* Local paths. These have the form

      <path>(\?<params)?

  where `<path>` must refer to (a subdirectory of) a Git repository.
  These differ from `file://` flake references in a few ways:

  * They refer to the working tree (unless an explicit `rev` or `ref`
    is specified), so evaluation can access dirty files. (Dirty files
    are files that are tracked by git but have uncommitted changes.)

  * The `dir` parameter is automatically derived. For example, if
    `/foo/bar` is a Git repository, then the flake reference
    `/foo/bar/flake` is equivalent to `/foo/bar?dir=flake`.

* GitHub repositories. These are downloaded as tarball archives,
  rather than through Git. This is often much faster and uses less
  disk space since it doesn't require fetching the entire history of
  the repository. On the other hand, it doesn't allow incremental
  fetching (but full downloads are often faster than incremental
  fetches!). The syntax is:

      github:<owner>/<repo>(/<rev-or-ref>)?(\?<params>)?

  `<rev-or-ref>` specifies the name of a branch or tag (`ref`), or a
  commit hash (`rev`). Note that unlike Git, GitHub allows fetching by

>   commit hash without specifying a branch or tag.

Reviewer 4: According to https://github.com/NixOS/nixos-channel-scripts/pull/27#issuecomment-547376718 flakes can’t handle slashes in branch names.

I think flakes not working with arbitrary branch names will be extremely annoying, and would be a mistake.

Author: Strictly speaking you can use slashes, but you have to write something like `?ref=refs/heads/channels/nixos-19.09`.

Reviewer 4: I suppose that’s fine, as long as it’s made extremely obvious to the user what they’ve done wrong if they try to use a slash and it doesn’t work.

Reviewer 4: Would also be useful to have `nix flake simplify-url` or something so that if you were generating a flake URI, you could just always use `ref`, full URLs, etc, and have a way of converting it to a simpler form if possible, rather than having to account for all these weird bits yourself.

But even then it’s starting to feel like quite a big ask for a human or computer to be able to construct these.

Reviewer 8: Looking at the RFC, it doesn't really seem clear *why* slashes in a `github:` uri aren't allowed? For local paths, it says the `dir` parameter is automatically derived; it doesn't read like this would be the case for git{,hub} uris.

    not available for `github` repositories (see below), since they're
    fetched as tarballs rather than as Git repositories.

  * `lastModifiedDate`: The commit time of the revision `rev`, in the
    format `%Y%m%d%H%M%S` (e.g. `20181231100934`). Unlike `revCount`,
    this is available for both Git and GitHub repositories, so it's
    useful for generating (hopefully) monotonically increasing version
    strings.

>   * `lastModified`: The commit time of the revision `rev` as an integer

Reviewer 51: Needs to be `lastModifiedDate`

Author: No, `lastModified` is an integer (seconds in the epoch) while `lastModifiedDate` is a string representation.

```
{
  edition = 201909;

  description = "A filesystem that fetches DWARF debug info from the Internet on demand";

>   outputs = { self, nixpkgs }: rec {

Reviewer 25: Should the set be recursive? Just read yesterday about how recursive overlays should be avoided and the proper way is to use the self fixpoint.

Author: No, it doesn't have to be recursive, you can also use `self` to refer to other outputs.

Reviewer 25: So the [advice about overlays](https://blog.flyingcircus.io/2017/11/07/nixos-the-dos-and-donts-of-nixpkgs-overlays/), specifically about a "recursive attrset", doesn't apply?

Author: Yes, because this is not an overlay and there is no override mechanism at the moment (so the `rec` is the same as `self.outputs`).

where getting flake metadata requires the execution of an arbitrarily
complex, possibly non-terminating program.

Flakes could be implemented as an external tool on top of Nix. Indeed,
there is nothing that flakes allow you to do that couldn't previously
be done using `fetchGit`, the `--pure-eval` flag and some shell
scripting. However, implementing flake-like functionality in an

> external tool would defeat the goals of this RFC. First, it probably

Reviewer 25: What's the reasoning behind Flakes deserving to be blessed as *the* standard before they've proven their merit or even having given other implementations a chance? Or is the intention to race RFCs in becoming the standard?

I like the Flakes RFC btw. I just don't see why it should be treated differently.

Author: > What's the reasoning behind Flakes deserving to be blessed as the standard before they've proven their merit or even having given other implementations a chance?

Well, no Nix feature has ever followed that process before being included, except `nix-repl` (and that was only because I didn't want an ncurses dependency in Nix at the time).

I'm very skeptical that external implementations would ever end up being merged in practice. See `yum`, which is still separate from `rpm` after 20 years.

Reviewer 4: Counterexamples: Cargo, `go mod`.

Reviewer 52: If anything, I think for a shift such as the one this RFC suggests, trying it out in practice somewhat at scale would make it *easier* to adopt it.  That way, potential issues in the design can be revealed and iterated on/addressed *before* we bolt the design down.

I'm not really familiar with the rpm ecosystem, so I don't know why `rpm` and `yum` are separate tools. It might in part just boil down to historical reasons though. 

I'm not sure I see why the (to me fairly marginal) risk of having two tools instead of one is bigger than the risk of the community moving in a direction and then looking back in retrospect and wanting to make changes to the flake design.

Don't get me wrong, I like a lot in the Flakes design. I just don't want us to have to repeat this process in a few years' time just because we didn't test the design well enough before giving it its blessing.

Reviewer 2: Given the current revision of this RFC to only talk about the flake format, this paragraph seems out of place.

Reviewer 4: I’m not sure it is. Whether it would be possible to generate this lockfile using an out-of-tree tool, or has to be implemented in Nix itself, is an important thing to know.

Reviewer 5: I don't know the full history, but I believe families like apt/deb, yum/rpm/dnf, gem/bundler were often built separately so as not to risk disruption by drastically modifying the (legacy) original tool, or to add on extra layers of functionality which only a subset of the community wanted (at the time).

If external-flakes is built with the explicit goal that it will be modified until it's either made part of nix or abandoned, that seems like it would alleviate that risk perfectly well.

I for one would love to try out a real implementation before comitting to it, since it's much harder to spot some usability or functionality gaps in an RFC than in practice. And I have a bunch of my own projects which are ripe for trying out a distributed approach to nix expressions.

Reviewer: > I'm very skeptical that external implementations would ever end up being merged in practice. See yum, which is still separate from rpm after 20 years.

Technically speaking, Nix and Nixpkgs are still separate repositories, and there are some cases where it causes a need for more complicated coordination. Still, Nixpkgs is the defacto repository of packages for Nix, but there are forks that are usable by anyone who is interested.

At the same time, NixOS _was_ a separate repository at some point, then got into the same monorepo. So merges happen here if they make sense.

What is the specific problem being addressed by flakes being built-in and not just the defacto tool?
